diff -crBN --binary sgl302/demos/chrome/main.c sgl302_patch/demos/chrome/main.c
*** sgl302/demos/chrome/main.c	1996-10-13 09:01:14.000000000 -0400
--- sgl302_patch/demos/chrome/main.c	2024-08-01 16:48:28.000000000 -0400
***************
*** 13,22 ****
  //	TARGET
  //	SEGA SATURN
  //
! //	REVISION 
  //	00/00/00
  //
! //	ORIGINAL FILE 
  //	s:\sgl_3\sgl\sample\chrome\main.c
  //
  //------------------------------------------------------------------------
--- 13,22 ----
  //	TARGET
  //	SEGA SATURN
  //
! //	REVISION
  //	00/00/00
  //
! //	ORIGINAL FILE
  //	s:\sgl_3\sgl\sample\chrome\main.c
  //
  //------------------------------------------------------------------------
***************
*** 64,70 ****
  	GRTBL( 13, 16, 16 ),
  	GRTBL( 14, 16, 16 ),
  	GRTBL( 15, 16, 16 ),
! 	GRTBL( 16, 16, 16 ),						  
  	GRTBL( 17, 16, 16 ),
  	GRTBL( 18, 16, 16 ),
  	GRTBL( 19, 16, 16 ),
--- 64,70 ----
  	GRTBL( 13, 16, 16 ),
  	GRTBL( 14, 16, 16 ),
  	GRTBL( 15, 16, 16 ),
! 	GRTBL( 16, 16, 16 ),
  	GRTBL( 17, 16, 16 ),
  	GRTBL( 18, 16, 16 ),
  	GRTBL( 19, 16, 16 ),
***************
*** 117,122 ****
--- 117,124 ----
  	0xC9AE,
  };
  
+ void	Texture_Load(void);
+ 
  void ss_main(void)
  {
  
***************
*** 130,136 ****
  		{toFIXED(0.0), toFIXED(0.0), toFIXED(0.0)};
  	ANGLE cm_ang[XYZ] =
  		{DEGtoANG(0.0), DEGtoANG(0.0), DEGtoANG(0.0) };
! 								   
  
  
  
--- 132,138 ----
  		{toFIXED(0.0), toFIXED(0.0), toFIXED(0.0)};
  	ANGLE cm_ang[XYZ] =
  		{DEGtoANG(0.0), DEGtoANG(0.0), DEGtoANG(0.0) };
! 
  
  
  
***************
*** 179,191 ****
  
  		// SPIN OBJECT
  		slRotZ(t_angZ);
! 		slRotY(t_angY);			   
  		slRotX(t_angX);
! 								 
  		t_angZ	+=	DEGtoANG(0.25);
  		t_angY	+=	DEGtoANG(1.0);
  		t_angX	+=	DEGtoANG(1.1);
! 	
  		// DRAW OBJECT
  		slPutPolygonX((XPDATA*)&xpdata_test,light);
  
--- 181,193 ----
  
  		// SPIN OBJECT
  		slRotZ(t_angZ);
! 		slRotY(t_angY);
  		slRotX(t_angX);
! 
  		t_angZ	+=	DEGtoANG(0.25);
  		t_angY	+=	DEGtoANG(1.0);
  		t_angX	+=	DEGtoANG(1.1);
! 
  		// DRAW OBJECT
  		slPutPolygonX((XPDATA*)&xpdata_test,light);
  
***************
*** 197,205 ****
  
  		slPrintHex(	slHex2Dec(slGetVCount())	, slLocate(16,2));	// Get vertical count position
  
! 		
  		slSynch();	// SYNC SGL
  
  	}
  }
- 
--- 199,206 ----
  
  		slPrintHex(	slHex2Dec(slGetVCount())	, slLocate(16,2));	// Get vertical count position
  
! 
  		slSynch();	// SYNC SGL
  
  	}
  }
diff -crBN --binary sgl302/demos/flying/actor.c sgl302_patch/demos/flying/actor.c
*** sgl302/demos/flying/actor.c	1996-08-07 09:05:04.000000000 -0400
--- sgl302_patch/demos/flying/actor.c	2024-08-02 21:12:58.000000000 -0400
***************
*** 25,30 ****
--- 25,32 ----
  #include "vector.h"
  #include "actor.h"
  
+ void AISparks( void );
+ 
  static int num_actors_executing;
  static int actor_tree_extant = FALSE;
  
***************
*** 48,54 ****
  	PROCESS* p = (PROCESS*)n;
  	List* l = &(((ACTOR*)v2)->process);
  
! 	LstUnlinkNode( l, n );	
  	ACTOR_MEM_FREE( p );
  	return( CONTINUE );
  }
--- 50,56 ----
  	PROCESS* p = (PROCESS*)n;
  	List* l = &(((ACTOR*)v2)->process);
  
! 	LstUnlinkNode( l, n );
  	ACTOR_MEM_FREE( p );
  	return( CONTINUE );
  }
***************
*** 58,65 ****
  	EOBJECT* o = (EOBJECT*)n;
  	List* l = &(((ACTOR*)v2)->object);
  
! 	LstUnlinkNode( l, n );	
! 	
  	ACTOR_MEM_FREE( o );	// deletes object and graphic
  	return( CONTINUE );
  }
--- 60,67 ----
  	EOBJECT* o = (EOBJECT*)n;
  	List* l = &(((ACTOR*)v2)->object);
  
! 	LstUnlinkNode( l, n );
! 
  	ACTOR_MEM_FREE( o );	// deletes object and graphic
  	return( CONTINUE );
  }
***************
*** 77,83 ****
  		// 'twould be nice if we could get to a Tree** from the node
  		// we're currently working on without having to search the tree.
  
! 		TreeDeleteByKey( &actor_tree, victim->tnode.key );		
  		actor_tree_count--;
  		ACTOR_MEM_FREE( victim );
  }
--- 79,85 ----
  		// 'twould be nice if we could get to a Tree** from the node
  		// we're currently working on without having to search the tree.
  
! 		TreeDeleteByKey( &actor_tree, victim->tnode.key );
  		actor_tree_count--;
  		ACTOR_MEM_FREE( victim );
  }
***************
*** 101,107 ****
  
  	switch ((int)message)
  	{
! 		case AMESS_DESTROY		:	
  		actor_kill = ACTOR_MEM_ALLOC( sizeof (Node) );
  		if (actor_kill != NULL)
  		{
--- 103,109 ----
  
  	switch ((int)message)
  	{
! 		case AMESS_DESTROY		:
  		actor_kill = ACTOR_MEM_ALLOC( sizeof (Node) );
  		if (actor_kill != NULL)
  		{
***************
*** 114,120 ****
  		}
  		break;
  
! 		case AMESS_RS_RUN		:	
  		case AMESS_RS_WAIT		:
  		case AMESS_RS_SLEEP		:
  		case AMESS_RS_SUSPEND	:
--- 116,122 ----
  		}
  		break;
  
! 		case AMESS_RS_RUN		:
  		case AMESS_RS_WAIT		:
  		case AMESS_RS_SLEEP		:
  		case AMESS_RS_SUSPEND	:
***************
*** 128,138 ****
  		case AMESS_COLLIDE_ATT:
  		AISparks();
  		break;
! 	
  		case AMESS_COLLIDE_DEF:
  		current->invuln = 15;
  		break;
! 		
  		default:
  		break;
  	}
--- 130,140 ----
  		case AMESS_COLLIDE_ATT:
  		AISparks();
  		break;
! 
  		case AMESS_COLLIDE_DEF:
  		current->invuln = 15;
  		break;
! 
  		default:
  		break;
  	}
***************
*** 141,147 ****
  int _ActBroadcastCallback( Tree** node, long v1, long* v2 )
  {
  	ACTOR* a = (ACTOR*)(*node);
! 	
  	ActSendMessage( a, v1, (long) v2 );
  	return( CONTINUE );
  }
--- 143,149 ----
  int _ActBroadcastCallback( Tree** node, long v1, long* v2 )
  {
  	ACTOR* a = (ACTOR*)(*node);
! 
  	ActSendMessage( a, v1, (long) v2 );
  	return( CONTINUE );
  }
***************
*** 170,181 ****
  
  		case AFLAG_RS_SLEEP:
  		if (mess > AMESS_PRIORITY_2)
! 		block = TRUE; 
  		break;
! 		
  		case AFLAG_RS_SUSPEND:
  		if (mess > AMESS_PRIORITY_1)
! 		block = TRUE; 
  		break;
  
  		default:
--- 172,183 ----
  
  		case AFLAG_RS_SLEEP:
  		if (mess > AMESS_PRIORITY_2)
! 		block = TRUE;
  		break;
! 
  		case AFLAG_RS_SUSPEND:
  		if (mess > AMESS_PRIORITY_1)
! 		block = TRUE;
  		break;
  
  		default:
***************
*** 221,254 ****
  	actor_tree_extant = FALSE;
  }
  
! // 
! // ACTOR* ActCreateActor( 
! // 	char* actor_name, 
! // 	ACTOR_FLAGS flags, 
! // 	short priority, 
  // 	void (*message_handler)(),
  // 	ACTOR* parent,
! // 	int scratch  ) 
! // 
  // Allocates and initializes an actor with no objects or processes.
  // priority is shifted left by 16 bits then incremented until a unique
  // ID is created to serve as the index tag in the actor tree.
! // scratch is the number of bytes of scratch space to allocate; the 
  // actor scratch pointer is set to a buffer of this size.
  // Once the actor is created, an AMESS_CREATE message is sent to that
! // actor - the message handler should be prepared for this, and will 
  // probably create the processes and objects for the actor.
  //
  // returns pointer to new actor or NULL
  //
  
! ACTOR* ActCreateActor( 
! 	char* actor_name, 
! 	ACTOR_FLAGS flags, 
! 	short priority, 
  	MESSAGE_HAND message_handler,
  	ACTOR* parent,
! 	int scratch  ) 
  {
  	Tree** scan;
  	ACTOR* new_actor;
--- 223,256 ----
  	actor_tree_extant = FALSE;
  }
  
! //
! // ACTOR* ActCreateActor(
! // 	char* actor_name,
! // 	ACTOR_FLAGS flags,
! // 	short priority,
  // 	void (*message_handler)(),
  // 	ACTOR* parent,
! // 	int scratch  )
! //
  // Allocates and initializes an actor with no objects or processes.
  // priority is shifted left by 16 bits then incremented until a unique
  // ID is created to serve as the index tag in the actor tree.
! // scratch is the number of bytes of scratch space to allocate; the
  // actor scratch pointer is set to a buffer of this size.
  // Once the actor is created, an AMESS_CREATE message is sent to that
! // actor - the message handler should be prepared for this, and will
  // probably create the processes and objects for the actor.
  //
  // returns pointer to new actor or NULL
  //
  
! ACTOR* ActCreateActor(
! 	char* actor_name,
! 	ACTOR_FLAGS flags,
! 	short priority,
  	MESSAGE_HAND message_handler,
  	ACTOR* parent,
! 	int scratch  )
  {
  	Tree** scan;
  	ACTOR* new_actor;
***************
*** 287,293 ****
  		new_actor->scratch.v = NULL;
  
  	new_actor->invuln = 0;
! 	
  	TreeAdd( &actor_tree, &(new_actor->tnode) );
  	actor_tree_count++;
  
--- 289,295 ----
  		new_actor->scratch.v = NULL;
  
  	new_actor->invuln = 0;
! 
  	TreeAdd( &actor_tree, &(new_actor->tnode) );
  	actor_tree_count++;
  
***************
*** 296,325 ****
  	return( new_actor );
  }
  
! // 
! // PROCESS* ActAddProcess( 
! // 	ACTOR* actor, 
! // 	void (*function)(), 
! // 	int delay, 
! // 	int autodelay, 
! // 	long priority )		
! // 
  // adds a process to the actor's process list.
! // delay is the number of ticks before it will execute (0-1 = ASAP); 
! // autodelay, if not 0, is the number of ticks to recycle 
  // the delay with
  // priority is the execution priority within the actor's processes
  // only
! // 
  // returns pointer to new process or NULL
  //
  
! PROCESS* ActAddProcess( 
! 	ACTOR* actor, 
  	PROCESS_FUNC function,
! 	int delay, 
! 	int autodelay, 
! 	long priority )		
  {
  	PROCESS* new_process;
  
--- 298,327 ----
  	return( new_actor );
  }
  
! //
! // PROCESS* ActAddProcess(
! // 	ACTOR* actor,
! // 	void (*function)(),
! // 	int delay,
! // 	int autodelay,
! // 	long priority )
! //
  // adds a process to the actor's process list.
! // delay is the number of ticks before it will execute (0-1 = ASAP);
! // autodelay, if not 0, is the number of ticks to recycle
  // the delay with
  // priority is the execution priority within the actor's processes
  // only
! //
  // returns pointer to new process or NULL
  //
  
! PROCESS* ActAddProcess(
! 	ACTOR* actor,
  	PROCESS_FUNC function,
! 	int delay,
! 	int autodelay,
! 	long priority )
  {
  	PROCESS* new_process;
  
***************
*** 336,348 ****
  		LstAddNodeByKey( &(actor->process), &(new_process->lnode) );
  	}
  	return( new_process );
! }		
  
  //
! // EOBJECT* ActAddObject( 
! //	ACTOR* actor, 
! //	OBJECT_FLAGS flags, 
! //	int num_verts, 
  //	int num_patches )
  //
  // adds an object with the specified flags
--- 338,350 ----
  		LstAddNodeByKey( &(actor->process), &(new_process->lnode) );
  	}
  	return( new_process );
! }
  
  //
! // EOBJECT* ActAddObject(
! //	ACTOR* actor,
! //	OBJECT_FLAGS flags,
! //	int num_verts,
  //	int num_patches )
  //
  // adds an object with the specified flags
***************
*** 353,374 ****
  // returns a pointer to new object or NULL
  //
  
! EOBJECT* ActAddObject( 
! 	ACTOR* actor, 
! 	OBJECT_FLAGS flags, 
! 	int num_verts, 
  	int num_patches )
  {
  	EOBJECT* new_object;
  	int size;
! 	
  	assert( actor != NULL );
! 	
  	switch	(flags & OFLAG_MASK_GRAPHIC)
  	{
  		case OFLAG_NVGRAPH:
! 		size =	sizeof( EOBJECT ) + 
! 				sizeof( NVGRAPH ) + 
  				(num_verts * sizeof( NVVERTEX )) +
  				(num_patches * sizeof( NVPATCH ));
  
--- 355,376 ----
  // returns a pointer to new object or NULL
  //
  
! EOBJECT* ActAddObject(
! 	ACTOR* actor,
! 	OBJECT_FLAGS flags,
! 	int num_verts,
  	int num_patches )
  {
  	EOBJECT* new_object;
  	int size;
! 
  	assert( actor != NULL );
! 
  	switch	(flags & OFLAG_MASK_GRAPHIC)
  	{
  		case OFLAG_NVGRAPH:
! 		size =	sizeof( EOBJECT ) +
! 				sizeof( NVGRAPH ) +
  				(num_verts * sizeof( NVVERTEX )) +
  				(num_patches * sizeof( NVPATCH ));
  
***************
*** 385,392 ****
  		break;
  
  		case OFLAG_SATPOLY:
! 		size =	sizeof( EOBJECT ) + 
! 				sizeof( SATPOLY ) + 
  				(num_verts * sizeof( POINT )) +
  				(num_patches * (sizeof( POLYGON )+sizeof( ATTR )));
  
--- 387,394 ----
  		break;
  
  		case OFLAG_SATPOLY:
! 		size =	sizeof( EOBJECT ) +
! 				sizeof( SATPOLY ) +
  				(num_verts * sizeof( POINT )) +
  				(num_patches * (sizeof( POLYGON )+sizeof( ATTR )));
  
***************
*** 403,409 ****
  		break;
  
  		case OFLAG_SATSPRITE:
! 		size =	sizeof( EOBJECT ) + 
  				sizeof( SATSPRITE );
  
  		new_object = ACTOR_MEM_ALLOC( size );
--- 405,411 ----
  		break;
  
  		case OFLAG_SATSPRITE:
! 		size =	sizeof( EOBJECT ) +
  				sizeof( SATSPRITE );
  
  		new_object = ACTOR_MEM_ALLOC( size );
***************
*** 416,422 ****
  		break;
  
  		case OFLAG_SATBACKGR:
! 		size =	sizeof( EOBJECT ) + 
  				sizeof( SATBACKGR );
  
  		new_object = ACTOR_MEM_ALLOC( size );
--- 418,424 ----
  		break;
  
  		case OFLAG_SATBACKGR:
! 		size =	sizeof( EOBJECT ) +
  				sizeof( SATBACKGR );
  
  		new_object = ACTOR_MEM_ALLOC( size );
***************
*** 444,450 ****
  int _ActUpdateProcessCallback( Node* n, long v1, long* v2 )
  {
  	PROCESS* p = (PROCESS*)n;
! 	
  	if (p->delay <= 1)
  	{
  		(*p->function)( (ACTOR*)v2, p );
--- 446,452 ----
  int _ActUpdateProcessCallback( Node* n, long v1, long* v2 )
  {
  	PROCESS* p = (PROCESS*)n;
! 
  	if (p->delay <= 1)
  	{
  		(*p->function)( (ACTOR*)v2, p );
***************
*** 517,523 ****
  	{
  	 	LstIterateValue( &(a->process), _ActUpdateProcessCallback, 0, (long*)a );
  		num_actors_executing++;
! 	}		
  
  	return( CONTINUE );
  }
--- 519,525 ----
  	{
  	 	LstIterateValue( &(a->process), _ActUpdateProcessCallback, 0, (long*)a );
  		num_actors_executing++;
! 	}
  
  	return( CONTINUE );
  }
diff -crBN --binary sgl302/demos/flying/actor.h sgl302_patch/demos/flying/actor.h
*** sgl302/demos/flying/actor.h	1996-08-07 13:53:02.000000000 -0400
--- sgl302_patch/demos/flying/actor.h	2024-08-05 16:23:48.000000000 -0400
***************
*** 28,33 ****
--- 28,35 ----
  
  //-------------------------- MACROS & TYPEDEFS ---------------------------
  
+ extern 	pointer	Memmalloc				(dword size);
+ 
  #define ACTOR_MEM_ALLOC(size)		Memmalloc(size)
  #define ACTOR_MEM_FREE(ptr)			Memfree(ptr)
  
***************
*** 50,56 ****
  #define DELAY_1HZ		60
  
  //--------------------------- GLOBAL VARIABLES ---------------------------
! 	
  extern char* act_xyz_order;
  extern char* act_xzy_order;
  extern char* act_yxz_order;
--- 52,58 ----
  #define DELAY_1HZ		60
  
  //--------------------------- GLOBAL VARIABLES ---------------------------
! 
  extern char* act_xyz_order;
  extern char* act_xzy_order;
  extern char* act_yxz_order;
***************
*** 63,85 ****
  extern void ActSendMessage( ACTOR* receive, MESSAGE_FLAGS mess, long data );
  extern void ActCreateActorTree( void );
  extern void ActKillActorTree( void );
! extern ACTOR* ActCreateActor( 
! 	char* actor_name, 
! 	ACTOR_FLAGS flags, 
! 	short priority, 
  	MESSAGE_HAND message_handler,
  	ACTOR* parent,
  	int scratch  );
! extern PROCESS* ActAddProcess( 
! 	ACTOR* actor, 
  	PROCESS_FUNC function,
! 	int delay, 
! 	int autodelay, 
! 	long priority );		
! extern EOBJECT* ActAddObject( 
! 	ACTOR* actor, 
! 	OBJECT_FLAGS flags, 
! 	int num_verts, 
  	int num_patches );
  extern int ActUpdate( long clock );
  extern void ProcSetDelay( PROCESS* proc, int delay, int autodelay );
--- 65,87 ----
  extern void ActSendMessage( ACTOR* receive, MESSAGE_FLAGS mess, long data );
  extern void ActCreateActorTree( void );
  extern void ActKillActorTree( void );
! extern ACTOR* ActCreateActor(
! 	char* actor_name,
! 	ACTOR_FLAGS flags,
! 	short priority,
  	MESSAGE_HAND message_handler,
  	ACTOR* parent,
  	int scratch  );
! extern PROCESS* ActAddProcess(
! 	ACTOR* actor,
  	PROCESS_FUNC function,
! 	int delay,
! 	int autodelay,
! 	long priority );
! extern EOBJECT* ActAddObject(
! 	ACTOR* actor,
! 	OBJECT_FLAGS flags,
! 	int num_verts,
  	int num_patches );
  extern int ActUpdate( long clock );
  extern void ProcSetDelay( PROCESS* proc, int delay, int autodelay );
***************
*** 89,92 ****
  extern EOBJECT* ActGetNextObject( ACTOR* act, EOBJECT* obj );
  
  #endif
- 
--- 91,93 ----
diff -crBN --binary sgl302/demos/flying/ai.c sgl302_patch/demos/flying/ai.c
*** sgl302/demos/flying/ai.c	1996-08-07 09:16:42.000000000 -0400
--- sgl302_patch/demos/flying/ai.c	2024-08-05 17:14:54.000000000 -0400
***************
*** 40,49 ****
  #define CONTROL_SAUCER	((actControl->scratch.l)[0])
  #define CONTROL_CAMERA	((actControl->scratch.l)[1])
  
! ATTR tile_attr = 
  	ATTRIBUTE(Dual_Plane, SORT_CEN, TEX_TILE, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
! SPR_ATTR spark_attr = 
  	SPR_ATTRIBUTE( TEX_SPARK16, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  
  #define SPARK_AGE ((act->scratch.l)[0])
--- 40,49 ----
  #define CONTROL_SAUCER	((actControl->scratch.l)[0])
  #define CONTROL_CAMERA	((actControl->scratch.l)[1])
  
! ATTR tile_attr =
  	ATTRIBUTE(Dual_Plane, SORT_CEN, TEX_TILE, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
! SPR_ATTR spark_attr =
  	SPR_ATTRIBUTE( TEX_SPARK16, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  
  #define SPARK_AGE ((act->scratch.l)[0])
***************
*** 51,57 ****
  void AISparkProcess1( ACTOR* act, PROCESS* proc )
  {
  	EOBJECT* obj;
- 	EOBJECT* dad;
  	SATSPRITE* ss;
  
  	obj = ActGetFirstObject( act );
--- 51,56 ----
***************
*** 67,73 ****
  void AISparkBuild( ACTOR* act )
  {
  	EOBJECT* obj;
- 	EOBJECT* dad;
  	SATSPRITE* ss;
  
  	obj = ActAddObject( act, OFLAG_SATSPRITE|OFLAG_AUTOMOVE|OFLAG_GRAVITY, 0, 0 );
--- 66,71 ----
***************
*** 127,137 ****
  	POINT* pt;
  	ATTR* at;
  
- 	int x, y, v;
- 	FIXED fx, fy;
- 
  	obj = ActGetFirstObject( act );
! 	
  	obj->pos.x = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
  	obj->pos.y = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
  	obj->pos.z = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
--- 125,132 ----
  	POINT* pt;
  	ATTR* at;
  
  	obj = ActGetFirstObject( act );
! 
  	obj->pos.x = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
  	obj->pos.y = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
  	obj->pos.z = FRANDOM( 1000.0 ) - toFIXED( 500.0 );
***************
*** 147,163 ****
  		}
  	}
  
! 	obj->rot.x = (ANGLE)slRandom(); 
! 	obj->rot.y = (ANGLE)slRandom(); 
! 	obj->rot.z = (ANGLE)slRandom(); 
! 	obj->rotvel.x = ((ANGLE)slRandom()) >> 5; 
! 	obj->rotvel.y = ((ANGLE)slRandom()) >> 5; 
! 	obj->rotvel.z = ((ANGLE)slRandom()) >> 5; 
  	obj->vel.x = toFIXED( 0.0 );
  	obj->vel.y = toFIXED( 0.0 );
  	obj->vel.z = toFIXED( 0.0 );
  	TILE_AGE = 0;
! 	
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
--- 142,158 ----
  		}
  	}
  
! 	obj->rot.x = (ANGLE)slRandom();
! 	obj->rot.y = (ANGLE)slRandom();
! 	obj->rot.z = (ANGLE)slRandom();
! 	obj->rotvel.x = ((ANGLE)slRandom()) >> 5;
! 	obj->rotvel.y = ((ANGLE)slRandom()) >> 5;
! 	obj->rotvel.z = ((ANGLE)slRandom()) >> 5;
  	obj->vel.x = toFIXED( 0.0 );
  	obj->vel.y = toFIXED( 0.0 );
  	obj->vel.z = toFIXED( 0.0 );
  	TILE_AGE = 0;
! 
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
***************
*** 178,184 ****
  
  	pt[3][X] = toFIXED( -50.0 );
  	pt[3][Y] = toFIXED(  50.0 );
! 	pt[3][Z] = toFIXED(   0.0 );			  
  
  	pg[0].norm[X] = toFIXED(  0.0 );
  	pg[0].norm[Y] = toFIXED(  0.0 );
--- 173,179 ----
  
  	pt[3][X] = toFIXED( -50.0 );
  	pt[3][Y] = toFIXED(  50.0 );
! 	pt[3][Z] = toFIXED(   0.0 );
  
  	pg[0].norm[X] = toFIXED(  0.0 );
  	pg[0].norm[Y] = toFIXED(  0.0 );
***************
*** 188,194 ****
  	pg[0].Vertices[2] = 2;
  	pg[0].Vertices[3] = 3;
  
! 	at[0] = tile_attr;	 
  
  	ProcSetFunction( proc, AITileProcess2 );
  }
--- 183,189 ----
  	pg[0].Vertices[2] = 2;
  	pg[0].Vertices[3] = 3;
  
! 	at[0] = tile_attr;
  
  	ProcSetFunction( proc, AITileProcess2 );
  }
***************
*** 331,337 ****
  	actMaster = AICreateActor( AT_KEY_MASTER, NULL );
  }
  
! ACTOR_TEMPLATE aiTemplates[] = 
  {
  	{	AT_KEY_MASTER,  	"Master",	AFLAG_RS_RUN, 10, AIMasterMsgHand,  8	},
  	{	AT_KEY_SAUCER,  	"Saucer",	AFLAG_RS_RUN, 20, AISaucerMsgHand,  32	},
--- 326,332 ----
  	actMaster = AICreateActor( AT_KEY_MASTER, NULL );
  }
  
! ACTOR_TEMPLATE aiTemplates[] =
  {
  	{	AT_KEY_MASTER,  	"Master",	AFLAG_RS_RUN, 10, AIMasterMsgHand,  8	},
  	{	AT_KEY_SAUCER,  	"Saucer",	AFLAG_RS_RUN, 20, AISaucerMsgHand,  32	},
***************
*** 355,364 ****
  	{
  		if (at->key == key)
  		{
! 			return( ActCreateActor( at->name, at->flags, at->pri, at->mh, parent, at->scr ) ); 
  		}
  		at++;
  	}
! 	return( NULL );		
  }
- 
--- 350,358 ----
  	{
  		if (at->key == key)
  		{
! 			return( ActCreateActor( at->name, at->flags, at->pri, at->mh, parent, at->scr ) );
  		}
  		at++;
  	}
! 	return( NULL );
  }
diff -crBN --binary sgl302/demos/flying/aictrl.c sgl302_patch/demos/flying/aictrl.c
*** sgl302/demos/flying/aictrl.c	1996-08-07 09:16:46.000000000 -0400
--- sgl302_patch/demos/flying/aictrl.c	2024-08-05 10:52:40.000000000 -0400
***************
*** 26,37 ****
  void AISaucerAntiGrav( ACTOR* act, FIXED v );
  void AISaucerAccelerate( ACTOR* act, FIXED x, FIXED y, FIXED z );
  void AISaucerShoot( ACTOR* act );
  
  #define CONTROL_SAUCER	((actControl->scratch.l)[0])
  #define CONTROL_CAMERA	((actControl->scratch.l)[1])
  
  #define PAD_START		PER_DGT_ST
! #define PAD_S			(PAD_START)	
  #define PAD_RIGHT		PER_DGT_KR
  #define PAD_R			(PAD_RIGHT)
  #define PAD_LEFT		PER_DGT_KL
--- 26,40 ----
  void AISaucerAntiGrav( ACTOR* act, FIXED v );
  void AISaucerAccelerate( ACTOR* act, FIXED x, FIXED y, FIXED z );
  void AISaucerShoot( ACTOR* act );
+ void AISaucerTurn( ACTOR* act, FIXED thrust );
+ void AISaucerSlip( ACTOR* act, FIXED thrust );
+ void AISaucerForward( ACTOR* act, FIXED thrust );
  
  #define CONTROL_SAUCER	((actControl->scratch.l)[0])
  #define CONTROL_CAMERA	((actControl->scratch.l)[1])
  
  #define PAD_START		PER_DGT_ST
! #define PAD_S			(PAD_START)
  #define PAD_RIGHT		PER_DGT_KR
  #define PAD_R			(PAD_RIGHT)
  #define PAD_LEFT		PER_DGT_KL
***************
*** 63,70 ****
  	if (Per_Connect1)
  	{
  		ctrlPadStatusCurrent = ~(Smpc_Peripheral[0].data);
! 	}	
! 	
  	ctrlPadStatusNew = ctrlPadStatusCurrent & (~ctrlPadStatusOld);
  	ctrlPadStatusDelta = ctrlPadStatusCurrent  ^ ctrlPadStatusOld;
  }
--- 66,73 ----
  	if (Per_Connect1)
  	{
  		ctrlPadStatusCurrent = ~(Smpc_Peripheral[0].data);
! 	}
! 
  	ctrlPadStatusNew = ctrlPadStatusCurrent & (~ctrlPadStatusOld);
  	ctrlPadStatusDelta = ctrlPadStatusCurrent  ^ ctrlPadStatusOld;
  }
***************
*** 105,108 ****
  		return( FALSE );
  	}
  }
- 
--- 108,110 ----
diff -crBN --binary sgl302/demos/flying/ailand.c sgl302_patch/demos/flying/ailand.c
*** sgl302/demos/flying/ailand.c	1996-08-07 09:16:50.000000000 -0400
--- sgl302_patch/demos/flying/ailand.c	2024-08-02 21:02:02.000000000 -0400
***************
*** 18,24 ****
  //------------------------------------------------------------------------
  #include <stdlib.h>
  
! // Floating point math package is used *only* to precalculate an 
  // interestingly curved "ground" surface. Not used during main run loop.
  
  #include <math.h>
--- 18,24 ----
  //------------------------------------------------------------------------
  #include <stdlib.h>
  
! // Floating point math package is used *only* to precalculate an
  // interestingly curved "ground" surface. Not used during main run loop.
  
  #include <math.h>
***************
*** 31,40 ****
  #include "enigma.h"
  #include "actor.h"
  #include "texture.h"
  
  #define random(range)		(rand()%(range))
  
! ATTR land_attr = 
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_LAND, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  #define LAND_X_PATCHES		6
--- 31,41 ----
  #include "enigma.h"
  #include "actor.h"
  #include "texture.h"
+ #include "vector.h"
  
  #define random(range)		(rand()%(range))
  
! ATTR land_attr =
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_LAND, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  #define LAND_X_PATCHES		6
***************
*** 67,82 ****
  	obj->pos.x = toFIXED(0.0);
  	obj->pos.y = LAND_Y_POS;
  	obj->pos.z = toFIXED(0.0);
! 	obj->rot.x = (ANGLE)0; 
! 	obj->rot.y = (ANGLE)0; 
! 	obj->rot.z = (ANGLE)0; 
! 	obj->rotvel.x = (ANGLE)0; 
! 	obj->rotvel.y = (ANGLE)0; 
! 	obj->rotvel.z = (ANGLE)0; 
  	obj->vel.x = toFIXED(0.0);
  	obj->vel.y = toFIXED(0.0);
  	obj->vel.z = toFIXED(0.0);
! 	
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
--- 68,83 ----
  	obj->pos.x = toFIXED(0.0);
  	obj->pos.y = LAND_Y_POS;
  	obj->pos.z = toFIXED(0.0);
! 	obj->rot.x = (ANGLE)0;
! 	obj->rot.y = (ANGLE)0;
! 	obj->rot.z = (ANGLE)0;
! 	obj->rotvel.x = (ANGLE)0;
! 	obj->rotvel.y = (ANGLE)0;
! 	obj->rotvel.z = (ANGLE)0;
  	obj->vel.x = toFIXED(0.0);
  	obj->vel.y = toFIXED(0.0);
  	obj->vel.z = toFIXED(0.0);
! 
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
***************
*** 89,97 ****
  		{
  			pt[v][X] = slMulFX( toFIXED(i-(LAND_X_PATCHES/2)), LAND_PATCH_SIZE );
  			pt[v][Z] = slMulFX( toFIXED(j-(LAND_Y_PATCHES/2)), LAND_PATCH_SIZE );
! 			
  			xx = (double)pt[v][X];
! 			xx /= 65536.0;				
  			xx /= 2.0;
  			zz = (double)pt[v][Z];
  			zz /= 65536.0;
--- 90,98 ----
  		{
  			pt[v][X] = slMulFX( toFIXED(i-(LAND_X_PATCHES/2)), LAND_PATCH_SIZE );
  			pt[v][Z] = slMulFX( toFIXED(j-(LAND_Y_PATCHES/2)), LAND_PATCH_SIZE );
! 
  			xx = (double)pt[v][X];
! 			xx /= 65536.0;
  			xx /= 2.0;
  			zz = (double)pt[v][Z];
  			zz /= 65536.0;
***************
*** 128,134 ****
  			pg[p].norm[Y] = norm.y;
  			pg[p].norm[Z] = norm.z;
  
! 			at[p] = land_attr;	 
  		}
  	}
  }
--- 129,135 ----
  			pg[p].norm[Y] = norm.y;
  			pg[p].norm[Z] = norm.z;
  
! 			at[p] = land_attr;
  		}
  	}
  }
***************
*** 145,148 ****
  		return( FALSE );
  	}
  }
- 
--- 146,148 ----
diff -crBN --binary sgl302/demos/flying/aisaucer.c sgl302_patch/demos/flying/aisaucer.c
*** sgl302/demos/flying/aisaucer.c	1996-08-07 09:16:54.000000000 -0400
--- sgl302_patch/demos/flying/aisaucer.c	2024-08-05 16:58:32.000000000 -0400
***************
*** 24,35 ****
  #include "texture.h"
  #include "assert.h"
  
! ATTR nose_attr = 
  	ATTRIBUTE(Dual_Plane, SORT_CEN, No_Texture, C_RGB(0, 31, 0), No_Gouraud, MESHoff, sprPolygon, UseClip );
! ATTR vect_attr = 
  	ATTRIBUTE(Dual_Plane, SORT_CEN, No_Texture, C_RGB(31, 0, 0), No_Gouraud, MESHoff, sprPolygon, UseClip );
  
! VECTOR3F dummy = { toFIXED(0.0), toFIXED(0.0), toFIXED(0.0) }; 
  
  void AIPointerProcess1( ACTOR* act, PROCESS* proc )
  {
--- 24,38 ----
  #include "texture.h"
  #include "assert.h"
  
! void VecSetMag( VECTOR3F* dest, FIXED scale );
! ANGLE VecAzimuth( VECTOR3F* a );
! 
! ATTR nose_attr =
  	ATTRIBUTE(Dual_Plane, SORT_CEN, No_Texture, C_RGB(0, 31, 0), No_Gouraud, MESHoff, sprPolygon, UseClip );
! ATTR vect_attr =
  	ATTRIBUTE(Dual_Plane, SORT_CEN, No_Texture, C_RGB(31, 0, 0), No_Gouraud, MESHoff, sprPolygon, UseClip );
  
! VECTOR3F dummy = { toFIXED(0.0), toFIXED(0.0), toFIXED(0.0) };
  
  void AIPointerProcess1( ACTOR* act, PROCESS* proc )
  {
***************
*** 58,64 ****
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pt = pdata->pntbl;
! 	
  	pt[1][X] = pt[2][X] = slMulFX( nose->x, toFIXED(200.0) );
  	pt[1][Y] = pt[2][Y] = slMulFX( nose->y, toFIXED(200.0) );
  	pt[1][Z] = pt[2][Z] = slMulFX( nose->z, toFIXED(200.0) );
--- 61,67 ----
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pt = pdata->pntbl;
! 
  	pt[1][X] = pt[2][X] = slMulFX( nose->x, toFIXED(200.0) );
  	pt[1][Y] = pt[2][Y] = slMulFX( nose->y, toFIXED(200.0) );
  	pt[1][Z] = pt[2][Z] = slMulFX( nose->z, toFIXED(200.0) );
***************
*** 71,101 ****
  void AIPointerBuild( ACTOR* act )
  {
  	EOBJECT* obj;
- 	EOBJECT* sobj;
  	SATPOLY* sg;
  	PDATA* pdata;
  	POLYGON* pg;
  	POINT* pt;
  	ATTR* at;
  
- 	int x, y, v;
- 	FIXED fx, fy;
- 
  	obj = ActAddObject( act, OFLAG_SATPOLY, 6, 2 );
! 	
  	obj->pos.x = toFIXED(0.0);
  	obj->pos.y = toFIXED(0.0);
  	obj->pos.z = toFIXED(0.0);
! 	obj->rot.x = 0; 
! 	obj->rot.y = 0; 
! 	obj->rot.z = 0; 
! 	obj->rotvel.x = (ANGLE)0; 
! 	obj->rotvel.y = (ANGLE)0; 
! 	obj->rotvel.z = (ANGLE)0; 
  	obj->vel.x = toFIXED( 0.0 );
  	obj->vel.y = toFIXED( 0.0 );
  	obj->vel.z = toFIXED( 0.0 );
! 	
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
--- 74,100 ----
  void AIPointerBuild( ACTOR* act )
  {
  	EOBJECT* obj;
  	SATPOLY* sg;
  	PDATA* pdata;
  	POLYGON* pg;
  	POINT* pt;
  	ATTR* at;
  
  	obj = ActAddObject( act, OFLAG_SATPOLY, 6, 2 );
! 
  	obj->pos.x = toFIXED(0.0);
  	obj->pos.y = toFIXED(0.0);
  	obj->pos.z = toFIXED(0.0);
! 	obj->rot.x = 0;
! 	obj->rot.y = 0;
! 	obj->rot.z = 0;
! 	obj->rotvel.x = (ANGLE)0;
! 	obj->rotvel.y = (ANGLE)0;
! 	obj->rotvel.z = (ANGLE)0;
  	obj->vel.x = toFIXED( 0.0 );
  	obj->vel.y = toFIXED( 0.0 );
  	obj->vel.z = toFIXED( 0.0 );
! 
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
***************
*** 112,118 ****
  
  	pt[3][X] = toFIXED(   0.0 );
  	pt[3][Y] = toFIXED( -20.0 );
! 	pt[3][Z] = toFIXED(   0.0 );			  
  
  	pg[0].norm[X] = toFIXED(  0.0 );
  	pg[0].norm[Y] = toFIXED(  0.0 );
--- 111,117 ----
  
  	pt[3][X] = toFIXED(   0.0 );
  	pt[3][Y] = toFIXED( -20.0 );
! 	pt[3][Z] = toFIXED(   0.0 );
  
  	pg[0].norm[X] = toFIXED(  0.0 );
  	pg[0].norm[Y] = toFIXED(  0.0 );
***************
*** 130,137 ****
  	pg[1].Vertices[2] = 5;
  	pg[1].Vertices[3] = 3;
  
! 	at[0] = nose_attr;	 
! 	at[1] = vect_attr;	 
  }
  
  int AIPointerMsgHand( ACTOR* act, MESSAGE_FLAGS mess, long data )
--- 129,136 ----
  	pg[1].Vertices[2] = 5;
  	pg[1].Vertices[3] = 3;
  
! 	at[0] = nose_attr;
! 	at[1] = vect_attr;
  }
  
  int AIPointerMsgHand( ACTOR* act, MESSAGE_FLAGS mess, long data )
***************
*** 150,163 ****
  
  VECTOR3F ai_blip_offset;
  
! SPR_ATTR blip_attr = 
  	SPR_ATTRIBUTE( TEX_BLIP, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  #define BLIP_AGE ((act->scratch.l)[0])
  
  void AIBlipProcess1( ACTOR* act, PROCESS* proc )
  {
  	EOBJECT* obj;
- 	EOBJECT* dad;
  	SATSPRITE* ss;
  
  	obj = ActGetFirstObject( act );
--- 149,161 ----
  
  VECTOR3F ai_blip_offset;
  
! SPR_ATTR blip_attr =
  	SPR_ATTRIBUTE( TEX_BLIP, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  #define BLIP_AGE ((act->scratch.l)[0])
  
  void AIBlipProcess1( ACTOR* act, PROCESS* proc )
  {
  	EOBJECT* obj;
  	SATSPRITE* ss;
  
  	obj = ActGetFirstObject( act );
***************
*** 191,197 ****
  	obj->vel.y = FRANDOM(8.0) - toFIXED(4.0);
  	obj->vel.z = FRANDOM(8.0) - toFIXED(4.0);
  
! 	obj->rotvel.z = (ANGLE)((slRandom()-toFIXED(0.5)) >> 3); 
  
  	BLIP_AGE = 0;
  }
--- 189,195 ----
  	obj->vel.y = FRANDOM(8.0) - toFIXED(4.0);
  	obj->vel.z = FRANDOM(8.0) - toFIXED(4.0);
  
! 	obj->rotvel.z = (ANGLE)((slRandom()-toFIXED(0.5)) >> 3);
  
  	BLIP_AGE = 0;
  }
***************
*** 210,216 ****
  	}
  }
  
! SPR_ATTR shot_attr = 
  	SPR_ATTRIBUTE( TEX_BALL, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  
  #define SHOT_AGE ((act->scratch.l)[0])
--- 208,214 ----
  	}
  }
  
! SPR_ATTR shot_attr =
  	SPR_ATTRIBUTE( TEX_BALL, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip );
  
  #define SHOT_AGE ((act->scratch.l)[0])
***************
*** 228,234 ****
  		ActSendMessage( act, AMESS_DESTROY, 0 );
  
  	ss->zrot += (ANGLE)20;
! 	
  }
  
  void AIShotBuild( ACTOR* act )
--- 226,232 ----
  		ActSendMessage( act, AMESS_DESTROY, 0 );
  
  	ss->zrot += (ANGLE)20;
! 
  }
  
  void AIShotBuild( ACTOR* act )
***************
*** 282,306 ****
  #define SAUCER_THROTTLE ((act->scratch.f)[1])
  #define SAUCER_FACING  ((VECTOR3F*)((act->scratch.f)+2))
  
! ATTR saucer_attr = 
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_SAUCER, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
! ATTR xsaucer_attr = 
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_SAUCER, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  void AISaucerBlips( ACTOR* act )
  {
- 	int i;
- 
  	if (slRandom() > toFIXED(0.7))
  		AICreateActor( AT_KEY_BLIP, act );
  }
  
  void AISaucerShoot( ACTOR* act )
  {
! 	ACTOR* shot;
! 
! 	shot = AICreateActor( AT_KEY_SHOT, act );
  }
  
  void AISaucerRestart( ACTOR* act )
--- 280,301 ----
  #define SAUCER_THROTTLE ((act->scratch.f)[1])
  #define SAUCER_FACING  ((VECTOR3F*)((act->scratch.f)+2))
  
! ATTR saucer_attr =
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_SAUCER, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
! ATTR xsaucer_attr =
  	ATTRIBUTE(Single_Plane, SORT_CEN, TEX_SAUCER, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  void AISaucerBlips( ACTOR* act )
  {
  	if (slRandom() > toFIXED(0.7))
  		AICreateActor( AT_KEY_BLIP, act );
  }
  
  void AISaucerShoot( ACTOR* act )
  {
! 	ACTOR* shot = AICreateActor( AT_KEY_SHOT, act );
! 	shot;
  }
  
  void AISaucerRestart( ACTOR* act )
***************
*** 335,343 ****
  	if (SAUCER_ANTIGRAV > toFIXED(40.0))
  		SAUCER_ANTIGRAV = toFIXED(40.0);
  
! 	ai_blip_offset.x = toFIXED(0.0); 
! 	ai_blip_offset.y = -slMulFX( v, toFIXED(30.0) ); 
! 	ai_blip_offset.z = toFIXED(0.0); 
  
  //	AISaucerBlips( act );
  }
--- 330,338 ----
  	if (SAUCER_ANTIGRAV > toFIXED(40.0))
  		SAUCER_ANTIGRAV = toFIXED(40.0);
  
! 	ai_blip_offset.x = toFIXED(0.0);
! 	ai_blip_offset.y = -slMulFX( v, toFIXED(30.0) );
! 	ai_blip_offset.z = toFIXED(0.0);
  
  //	AISaucerBlips( act );
  }
***************
*** 352,364 ****
  	if (SAUCER_THROTTLE < toFIXED(-3.0))
  		SAUCER_THROTTLE = toFIXED(-3.0);
  
! 	ai_blip_offset.x = -slMulFX( SAUCER_FACING->x, toFIXED(30.0) ); 
! 	ai_blip_offset.y = -slMulFX( SAUCER_FACING->y, toFIXED(30.0) ); 
! 	ai_blip_offset.z = -slMulFX( SAUCER_FACING->z, toFIXED(30.0) ); 
  
  //	AISaucerBlips( act );
  }
! 												  
  void AISaucerSlip( ACTOR* act, FIXED thrust )
  {
  	VECTOR3F up, right;
--- 347,359 ----
  	if (SAUCER_THROTTLE < toFIXED(-3.0))
  		SAUCER_THROTTLE = toFIXED(-3.0);
  
! 	ai_blip_offset.x = -slMulFX( SAUCER_FACING->x, toFIXED(30.0) );
! 	ai_blip_offset.y = -slMulFX( SAUCER_FACING->y, toFIXED(30.0) );
! 	ai_blip_offset.z = -slMulFX( SAUCER_FACING->z, toFIXED(30.0) );
  
  //	AISaucerBlips( act );
  }
! 
  void AISaucerSlip( ACTOR* act, FIXED thrust )
  {
  	VECTOR3F up, right;
***************
*** 369,385 ****
  	up.z = toFIXED(0.0);
  	VecCross( &right, SAUCER_FACING, &up );
  	VecSetMag( &right, thrust );
! 	
! 	ai_blip_offset.x = -slMulFX( right.x, toFIXED(30.0) ); 
! 	ai_blip_offset.y = -slMulFX( right.y, toFIXED(30.0) ); 
! 	ai_blip_offset.z = -slMulFX( right.z, toFIXED(30.0) ); 
  
  //	AISaucerBlips( act );
  
  	obj->vel.x += right.x;
  	assert( right.y == toFIXED( 0.0 ) );
  	obj->vel.z += right.z;
! 	
  	if (thrust > toFIXED(0.0))
  	{
  		if ((obj->rot.x > 0xE000u) || (obj->rot.x < 0x2000u))
--- 364,380 ----
  	up.z = toFIXED(0.0);
  	VecCross( &right, SAUCER_FACING, &up );
  	VecSetMag( &right, thrust );
! 
! 	ai_blip_offset.x = -slMulFX( right.x, toFIXED(30.0) );
! 	ai_blip_offset.y = -slMulFX( right.y, toFIXED(30.0) );
! 	ai_blip_offset.z = -slMulFX( right.z, toFIXED(30.0) );
  
  //	AISaucerBlips( act );
  
  	obj->vel.x += right.x;
  	assert( right.y == toFIXED( 0.0 ) );
  	obj->vel.z += right.z;
! 
  	if (thrust > toFIXED(0.0))
  	{
  		if ((obj->rot.x > 0xE000u) || (obj->rot.x < 0x2000u))
***************
*** 437,443 ****
  	45,  36,  27,   0,  -18,  -36,  -40
  };
  
! int vx_coords[ VERTEX_LAYERS ] = 
  {
  	 1,  27,  40, 114,  40,  27,   1
  };
--- 432,438 ----
  	45,  36,  27,   0,  -18,  -36,  -40
  };
  
! int vx_coords[ VERTEX_LAYERS ] =
  {
  	 1,  27,  40, 114,  40,  27,   1
  };
***************
*** 507,516 ****
  			pg[patch].norm[Z] = norm.z;
  
  			if (sector == 0)
! 				at[patch] = xsaucer_attr;	 
  			else
! 				at[patch] = saucer_attr;	 
! 		
  		}
  	}
  
--- 502,511 ----
  			pg[patch].norm[Z] = norm.z;
  
  			if (sector == 0)
! 				at[patch] = xsaucer_attr;
  			else
! 				at[patch] = saucer_attr;
! 
  		}
  	}
  
***************
*** 522,533 ****
  
  void AISaucerCruise( ACTOR* act, FIXED drag, FIXED stat )
  {
- 	ANGLE face_az;
- 	ANGLE move_az;
  	FIXED v;
  
  	EOBJECT* obj = ActGetFirstObject( act );
! 	
  	obj->vel.x += slMulFX( SAUCER_FACING->x, SAUCER_THROTTLE );
  	obj->vel.y += slMulFX( SAUCER_FACING->y, SAUCER_THROTTLE );
  	obj->vel.z += slMulFX( SAUCER_FACING->z, SAUCER_THROTTLE );
--- 517,526 ----
  
  void AISaucerCruise( ACTOR* act, FIXED drag, FIXED stat )
  {
  	FIXED v;
  
  	EOBJECT* obj = ActGetFirstObject( act );
! 
  	obj->vel.x += slMulFX( SAUCER_FACING->x, SAUCER_THROTTLE );
  	obj->vel.y += slMulFX( SAUCER_FACING->y, SAUCER_THROTTLE );
  	obj->vel.z += slMulFX( SAUCER_FACING->z, SAUCER_THROTTLE );
***************
*** 542,548 ****
  		v -= stat;
  	else
  		v = toFIXED( 0.0 );
! 	
  	VecSetMag( &(obj->vel), v );
  
  	slMulFX( SAUCER_THROTTLE, toFIXED( 0.9 ) );
--- 535,541 ----
  		v -= stat;
  	else
  		v = toFIXED( 0.0 );
! 
  	VecSetMag( &(obj->vel), v );
  
  	slMulFX( SAUCER_THROTTLE, toFIXED( 0.9 ) );
***************
*** 574,582 ****
  	if (alt < toFIXED(5.0))
  		alt = toFIXED(5.0);
  
! 	AISaucerAccelerate( act, toFIXED(0.0), slDivFX( slSquartFX(alt), SAUCER_ANTIGRAV ), toFIXED(0.0) ); 
  	AISaucerCruise( act, toFIXED(0.9), toFIXED(0.01) );
! 	
  }
  
  void AISaucerProcess1( ACTOR* act, PROCESS* proc )
--- 567,575 ----
  	if (alt < toFIXED(5.0))
  		alt = toFIXED(5.0);
  
! 	AISaucerAccelerate( act, toFIXED(0.0), slDivFX( slSquartFX(alt), SAUCER_ANTIGRAV ), toFIXED(0.0) );
  	AISaucerCruise( act, toFIXED(0.9), toFIXED(0.01) );
! 
  }
  
  void AISaucerProcess1( ACTOR* act, PROCESS* proc )
***************
*** 600,603 ****
  
  	}
  }
- 
--- 593,595 ----
diff -crBN --binary sgl302/demos/flying/aitarget.c sgl302_patch/demos/flying/aitarget.c
*** sgl302/demos/flying/aitarget.c	1996-08-09 06:54:18.000000000 -0400
--- sgl302_patch/demos/flying/aitarget.c	2024-08-05 16:53:14.000000000 -0400
***************
*** 24,30 ****
  #include "texture.h"
  #include "assert.h"
  
! ATTR target_attr = 
  	ATTRIBUTE(Dual_Plane, SORT_CEN, TEX_TARGET, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  void AITargetProcess1( ACTOR* act, PROCESS* proc )
--- 24,30 ----
  #include "texture.h"
  #include "assert.h"
  
! ATTR target_attr =
  	ATTRIBUTE(Dual_Plane, SORT_CEN, TEX_TARGET, No_Palet, No_Gouraud, CL32KRGB|MESHoff, sprNoflip, UseClip);
  
  void AITargetProcess1( ACTOR* act, PROCESS* proc )
***************
*** 34,40 ****
  void AITargetBuild( ACTOR* act )
  {
  	EOBJECT* obj;
- 	EOBJECT* sobj;
  	SATPOLY* sg;
  	PDATA* pdata;
  	POLYGON* pg;
--- 34,39 ----
***************
*** 43,62 ****
  	int v, f;
  
  	obj = ActAddObject( act, OFLAG_AUTOMOVE|OFLAG_AUTOROTATE|OFLAG_SATPOLY, 8, 6 );
! 	
  	obj->pos.x = FRANDOM( 4000.0 ) - toFIXED(2000.0);
  	obj->pos.y = FRANDOM( 300.0 ) + toFIXED( 100.0 );
  	obj->pos.z = FRANDOM( 4000.0 ) + toFIXED(200.0);
! 	obj->rot.x = (ANGLE)0; 
! 	obj->rot.y = (ANGLE)0; 
! 	obj->rot.z = (ANGLE)0; 
! 	obj->rotvel.x = (ANGLE)0; 
! 	obj->rotvel.y = (ANGLE)0; 
! 	obj->rotvel.z = (ANGLE)0; 
  	obj->vel.x = toFIXED(0.0);
  	obj->vel.y = toFIXED(0.0);
  	obj->vel.z = toFIXED(0.0);
! 	
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
--- 42,61 ----
  	int v, f;
  
  	obj = ActAddObject( act, OFLAG_AUTOMOVE|OFLAG_AUTOROTATE|OFLAG_SATPOLY, 8, 6 );
! 
  	obj->pos.x = FRANDOM( 4000.0 ) - toFIXED(2000.0);
  	obj->pos.y = FRANDOM( 300.0 ) + toFIXED( 100.0 );
  	obj->pos.z = FRANDOM( 4000.0 ) + toFIXED(200.0);
! 	obj->rot.x = (ANGLE)0;
! 	obj->rot.y = (ANGLE)0;
! 	obj->rot.z = (ANGLE)0;
! 	obj->rotvel.x = (ANGLE)0;
! 	obj->rotvel.y = (ANGLE)0;
! 	obj->rotvel.z = (ANGLE)0;
  	obj->vel.x = toFIXED(0.0);
  	obj->vel.y = toFIXED(0.0);
  	obj->vel.z = toFIXED(0.0);
! 
  	sg = (SATPOLY*)(obj->graph);
  	pdata = &(sg->pdata);
  	pg = pdata->pltbl;
***************
*** 75,81 ****
  		pg[f].norm[X] = toFIXED(  0.0 );
  		pg[f].norm[Y] = toFIXED(  0.0 );
  		pg[f].norm[Z] = toFIXED( -1.0 );
! 		at[f] = target_attr;	 
  	}
  
  	pg[0].Vertices[0] = 0;
--- 74,80 ----
  		pg[f].norm[X] = toFIXED(  0.0 );
  		pg[f].norm[Y] = toFIXED(  0.0 );
  		pg[f].norm[Z] = toFIXED( -1.0 );
! 		at[f] = target_attr;
  	}
  
  	pg[0].Vertices[0] = 0;
***************
*** 117,124 ****
  
  int AITargetMsgHand( ACTOR* act, MESSAGE_FLAGS mess, long data )
  {
- 	EOBJECT* obj;
- 
  	switch( mess )
  	{
  		case AMESS_CREATE:
--- 116,121 ----
***************
*** 137,140 ****
  		return( FALSE );
  	}
  }
- 
--- 134,136 ----
diff -crBN --binary sgl302/demos/flying/assert.h sgl302_patch/demos/flying/assert.h
*** sgl302/demos/flying/assert.h	1996-08-07 13:54:04.000000000 -0400
--- sgl302_patch/demos/flying/assert.h	2024-08-02 20:51:12.000000000 -0400
***************
*** 22,27 ****
--- 22,28 ----
  #ifdef NDEBUG
  	#define assert( test )		((void)0)
  #else
+ 	void _sassert( char* message1, char* message2 );
  	void _assert( char* message );
  
  	#define _STR(x)	_VAL(x)
diff -crBN --binary sgl302/demos/flying/collide.c sgl302_patch/demos/flying/collide.c
*** sgl302/demos/flying/collide.c	1996-08-07 09:17:20.000000000 -0400
--- sgl302_patch/demos/flying/collide.c	2024-08-05 16:54:54.000000000 -0400
***************
*** 20,25 ****
--- 20,26 ----
  #include "enigma.h"
  #include "collide.h"
  #include "assert.h"
+ #include "actor.h"
  
  ACTOR** collide_alist;
  ACTOR** collide_dlist;
***************
*** 51,57 ****
  	if ((a->flags & AFLAG_DEFEND_MASK) && (a->invuln == 0))
  		collide_dlist[ collide_dlist_count++ ] = a;
  
! 	return( CONTINUE );	
  }
  
  void CollideBuildLists( void )
--- 52,58 ----
  	if ((a->flags & AFLAG_DEFEND_MASK) && (a->invuln == 0))
  		collide_dlist[ collide_dlist_count++ ] = a;
  
! 	return( CONTINUE );
  }
  
  void CollideBuildLists( void )
***************
*** 93,99 ****
  			if (dz > dx)
  				SWAP( dx, dz );
  
! 			dst = dx + (dy+dz>>2);
  		}
  
  		if (dst < (o->radius+collide_attack_object->radius))
--- 94,100 ----
  			if (dz > dx)
  				SWAP( dx, dz );
  
! 			dst = dx + ((dy+dz)>>2);
  		}
  
  		if (dst < (o->radius+collide_attack_object->radius))
***************
*** 104,113 ****
  			collide_loc.y = (o->pos.y + collide_attack_object->pos.y)/2;
  			collide_loc.z = (o->pos.z + collide_attack_object->pos.z)/2;
  
! 			ActSendMessage( d, AMESS_COLLIDE_DEF, a );
! 			ActSendMessage( a, AMESS_COLLIDE_ATT, d );
  		}
  	}
  }
  
  int _CollideAttackCheckCallback( Node* n, long v1, long* v2 )
--- 105,115 ----
  			collide_loc.y = (o->pos.y + collide_attack_object->pos.y)/2;
  			collide_loc.z = (o->pos.z + collide_attack_object->pos.z)/2;
  
! 			ActSendMessage( d, AMESS_COLLIDE_DEF, (long)a );
! 			ActSendMessage( a, AMESS_COLLIDE_ATT, (long)d );
  		}
  	}
+ 	return 0;
  }
  
  int _CollideAttackCheckCallback( Node* n, long v1, long* v2 )
***************
*** 120,127 ****
  	{
  		collide_attack_object = o;
  		LstIterateValue( &(d->object), _CollideDefendCheckCallback, (long)a, (long*)d );
! 	}	
! 	
  	return( CONTINUE );
  }
  
--- 122,129 ----
  	{
  		collide_attack_object = o;
  		LstIterateValue( &(d->object), _CollideDefendCheckCallback, (long)a, (long*)d );
! 	}
! 
  	return( CONTINUE );
  }
  
diff -crBN --binary sgl302/demos/flying/collide.h sgl302_patch/demos/flying/collide.h
*** sgl302/demos/flying/collide.h	1996-08-07 09:34:26.000000000 -0400
--- sgl302_patch/demos/flying/collide.h	2024-08-05 16:29:14.000000000 -0400
***************
*** 17,22 ****
--- 17,26 ----
  //
  //------------------------------------------------------------------------
  
+ #include "mem_mgr.h"
+ 
+ extern 	pointer	Memmalloc				(dword size);
+ 
  #define COLLIDE_MEM_ALLOC(size)			Memmalloc(size)
  #define COLLIDE_MEM_FREE(ptr)			Memfree(ptr)
  
diff -crBN --binary sgl302/demos/flying/master.h sgl302_patch/demos/flying/master.h
*** sgl302/demos/flying/master.h	1996-08-07 14:00:34.000000000 -0400
--- sgl302_patch/demos/flying/master.h	2024-08-05 16:40:30.000000000 -0400
***************
*** 55,60 ****
  #undef NDEBUG
  //#define NDEBUG			1
  
- #pragma warning ( disable : 4245 )
- 
  #endif
--- 55,58 ----
diff -crBN --binary sgl302/demos/flying/satrend.c sgl302_patch/demos/flying/satrend.c
*** sgl302/demos/flying/satrend.c	1996-08-07 09:28:34.000000000 -0400
--- sgl302_patch/demos/flying/satrend.c	2024-08-05 14:02:32.000000000 -0400
***************
*** 23,28 ****
--- 23,29 ----
  #include "vector.h"
  #include "texture.h"
  #include "assert.h"
+ #include "saturn.h"
  
  Tree* graphic_tree;
  EOBJECT* rendCamera = NULL;
***************
*** 77,88 ****
  			case 'z':	slRotZ( obj->rot.z );	break;
  			default:	assert( TRUE==FALSE );	break;
  		}
! 	}	
  
  	slPutPolygon( sgp );
  
  	slPopMatrix();
! }	
  
  void RendSatSpriteObject( EOBJECT* obj )
  {
--- 78,89 ----
  			case 'z':	slRotZ( obj->rot.z );	break;
  			default:	assert( TRUE==FALSE );	break;
  		}
! 	}
  
  	slPutPolygon( sgp );
  
  	slPopMatrix();
! }
  
  void RendSatSpriteObject( EOBJECT* obj )
  {
***************
*** 101,112 ****
  	slPutSprite( pos, ss->attr, ss->zrot );
  
  	slPopMatrix();
! }	
  
  void RendSatBackObject( EOBJECT* obj )
  {
  
! }	
  
  int RendObjectCallback( Node* n )
  {
--- 102,113 ----
  	slPutSprite( pos, ss->attr, ss->zrot );
  
  	slPopMatrix();
! }
  
  void RendSatBackObject( EOBJECT* obj )
  {
  
! }
  
  int RendObjectCallback( Node* n )
  {
***************
*** 126,132 ****
  	dst = dx + (dy+dz>>2);
  
  	if (dst > rendGrossDistanceClip)
! 		return( CONTINUE );	
  
  	switch( o->flags & OFLAG_MASK_GRAPHIC )
  	{
--- 127,133 ----
  	dst = dx + (dy+dz>>2);
  
  	if (dst > rendGrossDistanceClip)
! 		return( CONTINUE );
  
  	switch( o->flags & OFLAG_MASK_GRAPHIC )
  	{
***************
*** 148,154 ****
  		break;
  	}
  
! 	return( CONTINUE );	
  }
  
  int RendActorCallback( Tree** node )
--- 149,155 ----
  		break;
  	}
  
! 	return( CONTINUE );
  }
  
  int RendActorCallback( Tree** node )
diff -crBN --binary sgl302/demos/flying/saturn.c sgl302_patch/demos/flying/saturn.c
*** sgl302/demos/flying/saturn.c	1996-08-07 14:00:32.000000000 -0400
--- sgl302_patch/demos/flying/saturn.c	2024-08-05 16:55:16.000000000 -0400
***************
*** 17,22 ****
--- 17,24 ----
  //
  //------------------------------------------------------------------------
  
+ #include <sl_def.h>
+ 
  #include "saturn.h"
  #include "assert.h"
  
***************
*** 24,32 ****
  
  void SatInitTime( void )
  {
! 	unsigned char tcsr;
! 
! 	*SH2_TCR = SH2_TCR_CKS1;	// Ì/128 clock - 200KHz?
  
  	// clear overflow
  	*SH2_TCSR = 0;
--- 26,32 ----
  
  void SatInitTime( void )
  {
! 	*SH2_TCR = SH2_TCR_CKS1;	// ÔøΩ/128 clock - 200KHz?
  
  	// clear overflow
  	*SH2_TCSR = 0;
***************
*** 58,64 ****
  /*
  	assert handler
  
! */				  
  
  void _sassert( char* message1, char* message2 )
  {
--- 58,64 ----
  /*
  	assert handler
  
! */
  
  void _sassert( char* message1, char* message2 )
  {
***************
*** 70,76 ****
  
  	asm("trapa	#12");
  }
- 
- 
- 		 
- 
--- 70,72 ----
diff -crBN --binary sgl302/sample/common/cinit.c sgl302_patch/sample/common/cinit.c
*** sgl302/sample/common/cinit.c	1996-09-10 23:00:00.000000000 -0400
--- sgl302_patch/sample/common/cinit.c	2022-09-15 21:12:48.000000000 -0400
***************
*** 14,20 ****
  /* */
  extern void ss_main( void );
  
! void	main( void )
  {
  	Uint8	*dst;
  	Uint32	i;
--- 14,20 ----
  /* */
  extern void ss_main( void );
  
! int	main( void )
  {
  	Uint8	*dst;
  	Uint32	i;
***************
*** 32,36 ****
  
  	/* Application Call */
  	ss_main();
  }
! 
\ No newline at end of file
--- 32,38 ----
  
  	/* Application Call */
  	ss_main();
+ 	
+ 	return 0;
  }
! 
diff -crBN --binary sgl302/sample/demo_d/main.c sgl302_patch/sample/demo_d/main.c
*** sgl302/sample/demo_d/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/demo_d/main.c	2024-08-02 20:55:14.000000000 -0400
***************
*** 1,3 ****
--- 1,5 ----
+ #include <string.h>
+ 
  #include	"sgl.h"
  #include	"ss_scrol.h"
  
***************
*** 207,213 ****
  		disp_poly();
  		disp_scroll();
  		slSynch();
! 	} 
  }
  
! 
\ No newline at end of file
--- 209,215 ----
  		disp_poly();
  		disp_scroll();
  		slSynch();
! 	}
  }
  
! 
diff -crBN --binary sgl302/sample/demogun/main.c sgl302_patch/sample/demogun/main.c
*** sgl302/sample/demogun/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/demogun/main.c	2024-08-02 21:07:24.000000000 -0400
***************
*** 11,16 ****
--- 11,18 ----
  extern Sint8 SynchCount;
  extern Uint16 VDP2_EXTEN;
  
+ Bool slCheckIntBackSet(void);
+ 
  static Uint8 GunCtrlFlag1P, GunCtrlFlag2P;
  
  static Uint8 getMegaDriveId(Uint8 pd1, Uint8 pd0)
***************
*** 154,157 ****
  	}
  }
  
! 
\ No newline at end of file
--- 156,159 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample/design1/main.c sgl302_patch/sample/design1/main.c
*** sgl302/sample/design1/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/design1/main.c	2024-08-05 15:59:52.000000000 -0400
***************
*** 3,8 ****
--- 3,9 ----
  /*----------------------------------------------------------------------*/
  #include	"sgl.h"
  #include	"sega_sys.h"
+ #include	<string.h>
  
  extern PDATA pdata_sample[];
  
***************
*** 147,150 ****
  	}
  }
  
! 
\ No newline at end of file
--- 148,151 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample/design2/main.c sgl302_patch/sample/design2/main.c
*** sgl302/sample/design2/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/design2/main.c	2024-08-11 11:56:54.000000000 -0400
***************
*** 4,9 ****
--- 4,11 ----
  #include	"sgl.h"
  #include	"sega_sys.h"
  
+ #include	<string.h>
+ 
  #define		max_texture		2
  
  extern PDATA pdata_sample[];
***************
*** 163,166 ****
  	}
  }
  
! 
\ No newline at end of file
--- 165,168 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample/design3/main.c sgl302_patch/sample/design3/main.c
*** sgl302/sample/design3/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/design3/main.c	2024-08-11 12:04:36.000000000 -0400
***************
*** 4,9 ****
--- 4,11 ----
  #include	"sgl.h"
  #include	"sega_sys.h"
  
+ #include	<string.h>
+ 
  #define		max_texture		2
  
  extern OBJECT	object_sample[];
***************
*** 193,196 ****
  	}
  }
  
! 
\ No newline at end of file
--- 195,198 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample/driving/main.c sgl302_patch/sample/driving/main.c
*** sgl302/sample/driving/main.c	1996-09-10 23:00:00.000000000 -0400
--- sgl302_patch/sample/driving/main.c	2024-08-11 12:30:48.000000000 -0400
***************
*** 11,16 ****
--- 11,19 ----
  
  #include	"sglcolli.h"
  
+ #include <math.h>
+ #include <stdlib.h>
+ 
  #define FLASH_SATURN_CD
  
  
***************
*** 153,159 ****
  	slMakeKtable((void *)RBG0_KTB_ADR);
  	slCharRbg0(COL_TYPE_256 , CHAR_SIZE_1x1);
  	slPageRbg0((void *)RBG0RB_CEL_ADR , 0 , PNB_1WORD|CN_12BIT);
! 	
  	slPlaneRA(PL_SIZE_1x1);
  	sl1MapRA((void *)RBG0RA_MAP_ADR);
  	slOverRA(0);
--- 156,162 ----
  	slMakeKtable((void *)RBG0_KTB_ADR);
  	slCharRbg0(COL_TYPE_256 , CHAR_SIZE_1x1);
  	slPageRbg0((void *)RBG0RB_CEL_ADR , 0 , PNB_1WORD|CN_12BIT);
! 
  	slPlaneRA(PL_SIZE_1x1);
  	sl1MapRA((void *)RBG0RA_MAP_ADR);
  	slOverRA(0);
***************
*** 161,167 ****
  	Cel2VRAM(tuti_cel , (void *)RBG0RA_CEL_ADR , 65536);
  	Map2VRAM(tuti_map , (void *)RBG0RA_MAP_ADR , 64 , 64 , 2 , 884);
  	Pal2CRAM(tuti_pal , (void *)RBG0RA_COL_ADR , 160);
! 	
  	slPlaneRB(PL_SIZE_1x1);
  	sl1MapRB((void *)RBG0RB_MAP_ADR);
  	slOverRB(0);
--- 164,170 ----
  	Cel2VRAM(tuti_cel , (void *)RBG0RA_CEL_ADR , 65536);
  	Map2VRAM(tuti_map , (void *)RBG0RA_MAP_ADR , 64 , 64 , 2 , 884);
  	Pal2CRAM(tuti_pal , (void *)RBG0RA_COL_ADR , 160);
! 
  	slPlaneRB(PL_SIZE_1x1);
  	sl1MapRB((void *)RBG0RB_MAP_ADR);
  	slOverRB(0);
***************
*** 169,178 ****
  	Cel2VRAM(sora_cel , (void *)RBG0RB_CEL_ADR , 28288);
  	Map2VRAM(sora_map , (void *)RBG0RB_MAP_ADR , 64 , 20 , 1 , 0);
  	Pal2CRAM(sora_pal , (void *)RBG0RB_COL_ADR , 256);
! 	
  	slRparaMode(K_CHANGE);
  	slBack1ColSet((void *)BACK_COL_ADR , 0);
! 	
  	slScrAutoDisp(NBG0ON | RBG0ON );
  }
  /*--------------------------------------------------------------------------*/
--- 172,181 ----
  	Cel2VRAM(sora_cel , (void *)RBG0RB_CEL_ADR , 28288);
  	Map2VRAM(sora_map , (void *)RBG0RB_MAP_ADR , 64 , 20 , 1 , 0);
  	Pal2CRAM(sora_pal , (void *)RBG0RB_COL_ADR , 256);
! 
  	slRparaMode(K_CHANGE);
  	slBack1ColSet((void *)BACK_COL_ADR , 0);
! 
  	slScrAutoDisp(NBG0ON | RBG0ON );
  }
  /*--------------------------------------------------------------------------*/
***************
*** 190,206 ****
  /*--------------------------------------------------------------------------*/
  	init_camera((void*)&sega3D);
  	init_object((void*)&sega3D);
! 	init_matrix(pos, ang, scl);	
  	init_main();
  	pos_move();
  	slZdspLevel(5);
  	slWindowClipLevel(50);
  /*--------------------------------------------------------------------------*/
! 	slInitEvent();							/*  ÉCÉxÉìÉgÇÃèâä˙âª		*/
! 	slSetEvent((void *)main_pad);			/*  ÉpÉbÉhì¸óÕä÷åW			*/
! 	slSetEvent((void *)main_came);			/*  ÉJÉÅÉâà íuéwíËä÷åW		*/
! 	slSetEvent((void *)main_Pmap);			/*  É|ÉäÉSÉìÉ}ÉbÉvï\é¶ä÷åW	*/
! 	slSetEvent((void *)main_map);			/*  ÇaÇfï\é¶ä÷åW	*/
  /*--------------------------------------------------------------------------*/
  	while(1)
  	{
--- 193,209 ----
  /*--------------------------------------------------------------------------*/
  	init_camera((void*)&sega3D);
  	init_object((void*)&sega3D);
! 	init_matrix(pos, ang, scl);
  	init_main();
  	pos_move();
  	slZdspLevel(5);
  	slWindowClipLevel(50);
  /*--------------------------------------------------------------------------*/
! 	slInitEvent();							/*  ÔøΩCÔøΩxÔøΩÔøΩÔøΩgÔøΩÃèÔøΩÔøΩÔøΩÔøΩÔøΩ		*/
! 	slSetEvent((void *)main_pad);			/*  ÔøΩpÔøΩbÔøΩhÔøΩÔøΩÔøΩÕä÷åW			*/
! 	slSetEvent((void *)main_came);			/*  ÔøΩJÔøΩÔøΩÔøΩÔøΩÔøΩ íuÔøΩwÔøΩÔøΩÔøΩ÷åW		*/
! 	slSetEvent((void *)main_Pmap);			/*  ÔøΩ|ÔøΩÔøΩÔøΩSÔøΩÔøΩÔøΩ}ÔøΩbÔøΩvÔøΩ\ÔøΩÔøΩÔøΩ÷åW	*/
! 	slSetEvent((void *)main_map);			/*  ÔøΩaÔøΩfÔøΩ\ÔøΩÔøΩÔøΩ÷åW	*/
  /*--------------------------------------------------------------------------*/
  	while(1)
  	{
***************
*** 209,225 ****
  			SYS_Exit(1);
  		}
  #endif
! 		slExecuteEvent();					/*	ÉCÉxÉìÉgÇÃé¿çs			*/
  		slPrintFX(TotalPolygons<<16, slLocate(30,4));
  		slPrintFX(DispPolygons<<16, slLocate(30,6));
  //		slPrintFX(pos[X],slLocate(3,10));
  //		slPrintFX(pos[Y],slLocate(3,11));
! 		slSynch();							/*	ÉVÉìÉN					*/
! 	} 
  /*--------------------------------------------------------------------------*/
  }
  /*--------------------------------------*/
! /*	ÉJÉÅÉâä÷åW							*/
  /*--------------------------------------*/
  void	main_came(EVENT *evptr)
  {
--- 212,228 ----
  			SYS_Exit(1);
  		}
  #endif
! 		slExecuteEvent();					/*	ÔøΩCÔøΩxÔøΩÔøΩÔøΩgÔøΩÃéÔøΩÔøΩs			*/
  		slPrintFX(TotalPolygons<<16, slLocate(30,4));
  		slPrintFX(DispPolygons<<16, slLocate(30,6));
  //		slPrintFX(pos[X],slLocate(3,10));
  //		slPrintFX(pos[Y],slLocate(3,11));
! 		slSynch();							/*	ÔøΩVÔøΩÔøΩÔøΩN					*/
! 	}
  /*--------------------------------------------------------------------------*/
  }
  /*--------------------------------------*/
! /*	ÔøΩJÔøΩÔøΩÔøΩÔøΩÔøΩ÷åW							*/
  /*--------------------------------------*/
  void	main_came(EVENT *evptr)
  {
***************
*** 227,233 ****
  	slLookAt(sega3D.cpos, sega3D.ctarget, sega3D.cangle[Z]);
  }
  /*--------------------------------------*/
! /*	É|ÉäÉSÉìÉ}ÉbÉvä÷åWÇÃï\é¶			*/
  /*--------------------------------------*/
  void	main_Pmap(EVENT *evptr)
  {
--- 230,236 ----
  	slLookAt(sega3D.cpos, sega3D.ctarget, sega3D.cangle[Z]);
  }
  /*--------------------------------------*/
! /*	ÔøΩ|ÔøΩÔøΩÔøΩSÔøΩÔøΩÔøΩ}ÔøΩbÔøΩvÔøΩ÷åWÔøΩÃï\ÔøΩÔøΩ			*/
  /*--------------------------------------*/
  void	main_Pmap(EVENT *evptr)
  {
***************
*** 241,247 ****
  	Sint32	my_xx,my_yy;
  	Sint32	X_MAX,X_MIN;
  	Sint32	Y_MAX,Y_MIN;
! 	
  	my_xx	=((Sint16)((pos[X]+256*slSin(ang[Z]))>>16))>>5;
  	my_yy	=((Sint16)((pos[Y]+256*slCos(ang[Z]))>>16))>>5;
  	MapAngX	= slCos(ang[Z])*256;
--- 244,250 ----
  	Sint32	my_xx,my_yy;
  	Sint32	X_MAX,X_MIN;
  	Sint32	Y_MAX,Y_MIN;
! 
  	my_xx	=((Sint16)((pos[X]+256*slSin(ang[Z]))>>16))>>5;
  	my_yy	=((Sint16)((pos[Y]+256*slCos(ang[Z]))>>16))>>5;
  	MapAngX	= slCos(ang[Z])*256;
***************
*** 258,264 ****
  		Y_MAX	=(my_yy>>2)+15;
  		if(Y_MIN < 0)Y_MIN=0;
  		if(Y_MAX > pol_map_y)Y_MAX=pol_map_y;
! 		
  		for(ax = X_MIN ; ax < X_MAX ; ax++)
  		{
  			for(ay = Y_MIN ; ay < Y_MAX ; ay++)
--- 261,267 ----
  		Y_MAX	=(my_yy>>2)+15;
  		if(Y_MIN < 0)Y_MIN=0;
  		if(Y_MAX > pol_map_y)Y_MAX=pol_map_y;
! 
  		for(ax = X_MIN ; ax < X_MAX ; ax++)
  		{
  			for(ay = Y_MIN ; ay < Y_MAX ; ay++)
***************
*** 286,292 ****
  	slPopMatrix();
  }
  /*--------------------------------------*/
! /*	ÇaÇfï\é¶ä÷åW						*/
  /*--------------------------------------*/
  void	main_map(EVENT *evptr)
  {
--- 289,295 ----
  	slPopMatrix();
  }
  /*--------------------------------------*/
! /*	ÔøΩaÔøΩfÔøΩ\ÔøΩÔøΩÔøΩ÷åW						*/
  /*--------------------------------------*/
  void	main_map(EVENT *evptr)
  {
***************
*** 309,315 ****
  	pos_move();
  }
  /*--------------------------------------*/
! /*	à íuï€ë∂							*/
  /*--------------------------------------*/
  void	pos_move()
  {
--- 312,318 ----
  	pos_move();
  }
  /*--------------------------------------*/
! /*	ÔøΩ íuÔøΩ€ëÔøΩ							*/
  /*--------------------------------------*/
  void	pos_move()
  {
***************
*** 321,327 ****
  	ang2[Z]	=ang[Z];
  }
  /*--------------------------------------*/
! /*	ÉpÉbÉhä÷åW							*/
  /*--------------------------------------*/
  void	main_pad(EVENT *evptr)
  {
--- 324,330 ----
  	ang2[Z]	=ang[Z];
  }
  /*--------------------------------------*/
! /*	ÔøΩpÔøΩbÔøΩhÔøΩ÷åW							*/
  /*--------------------------------------*/
  void	main_pad(EVENT *evptr)
  {
***************
*** 331,337 ****
  	Sint16		aa;
  	Uint16		data;
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[15].data;					//ÉpÉbÉhÇQ
  	if((data & PER_DGT_TC) == 0)tak_data+=10;
  	if((data & PER_DGT_TZ) == 0)tak_data-=10;
  	if((data & PER_DGT_KL) == 0)ang[Z]+=DEGtoANG(4.0);
--- 334,340 ----
  	Sint16		aa;
  	Uint16		data;
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[15].data;					//ÔøΩpÔøΩbÔøΩhÔøΩQ
  	if((data & PER_DGT_TC) == 0)tak_data+=10;
  	if((data & PER_DGT_TZ) == 0)tak_data-=10;
  	if((data & PER_DGT_KL) == 0)ang[Z]+=DEGtoANG(4.0);
***************
*** 340,346 ****
  	if((data & PER_DGT_KD) == 0)ang[X]-=DEGtoANG(4.0);
  	ang[X]	=ang_z;
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[0].data;						//ÉpÉbÉhÇP
  #ifdef FLASH_SATURN_CD
  	if((data & (PER_DGT_ST | PER_DGT_TA | PER_DGT_TB | PER_DGT_TC)) == 0){
  		SYS_Exit(0);
--- 343,349 ----
  	if((data & PER_DGT_KD) == 0)ang[X]-=DEGtoANG(4.0);
  	ang[X]	=ang_z;
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[0].data;						//ÔøΩpÔøΩbÔøΩhÔøΩP
  #ifdef FLASH_SATURN_CD
  	if((data & (PER_DGT_ST | PER_DGT_TA | PER_DGT_TB | PER_DGT_TC)) == 0){
  		SYS_Exit(0);
***************
*** 348,356 ****
  #endif
  	if((data & PER_DGT_ST) == 0)init_main();
  /*---------------------------------------------*/
! /*é‘ÇÃãììÆÇÕÉAÉoÉEÉgÇ…çÏÇ¡ÇƒÇ¢ÇÈÇÃÇ≈éQçlÇ…ÇµÇ»Ç¢Ç≈Ç≠ÇæÇ≥Ç¢ÅB*/
  /*---------------------------------------------*/
! 	/*â¡ë¨å∏ë¨*/
  	aa	=0;
  	if((data & PER_DGT_TB) == 0)
  	{
--- 351,359 ----
  #endif
  	if((data & PER_DGT_ST) == 0)init_main();
  /*---------------------------------------------*/
! /*ÔøΩ‘ÇÃãÔøΩÔøΩÔøΩÔøΩÕÉAÔøΩoÔøΩEÔøΩgÔøΩ…çÔøΩÔøΩÔøΩÔøΩƒÇÔøΩÔøΩÔøΩÔøΩÃÇ≈éQÔøΩlÔøΩ…ÇÔøΩÔøΩ»ÇÔøΩÔøΩ≈ÇÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩB*/
  /*---------------------------------------------*/
! 	/*ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩ*/
  	aa	=0;
  	if((data & PER_DGT_TB) == 0)
  	{
***************
*** 372,378 ****
  	pos_add[X]	-=eg_speed*slSin(ang[Z])/10+col_x;
  	pos_add[Y]	-=eg_speed*slCos(ang[Z])/10+col_y;
  	/*---------------------------------------------*/
! 	/*ÉTÉCÉhà⁄ìÆ*/
  	if((data & PER_DGT_TL) == 0)
  	{
  		pos[X]	+=3*slCos(ang[Z]);
--- 375,381 ----
  	pos_add[X]	-=eg_speed*slSin(ang[Z])/10+col_x;
  	pos_add[Y]	-=eg_speed*slCos(ang[Z])/10+col_y;
  	/*---------------------------------------------*/
! 	/*ÔøΩTÔøΩCÔøΩhÔøΩ⁄ìÔøΩ*/
  	if((data & PER_DGT_TL) == 0)
  	{
  		pos[X]	+=3*slCos(ang[Z]);
***************
*** 383,389 ****
  		pos[Y]	+=3*slSin(ang[Z]);
  	}
  	/*---------------------------------------------*/
! 	/*ç∂âEÇÃâÒì]*/
  	aa	=0;
  	if((data & PER_DGT_KR) == 0)
  	{
--- 386,392 ----
  		pos[Y]	+=3*slSin(ang[Z]);
  	}
  	/*---------------------------------------------*/
! 	/*ÔøΩÔøΩÔøΩEÔøΩÃâÔøΩÔøΩ]*/
  	aa	=0;
  	if((data & PER_DGT_KR) == 0)
  	{
***************
*** 432,438 ****
  	by	=ay-30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
  	point_l	=col_att;
! 	
  	bx	=ax-30*slCos(ang[Z]);
  	by	=ay+30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
--- 435,441 ----
  	by	=ay-30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
  	point_l	=col_att;
! 
  	bx	=ax-30*slCos(ang[Z]);
  	by	=ay+30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
***************
*** 489,495 ****
  	/*---------------------------------------------*/
  }
  /*-------------------------------------------------*/
! /*	ÉRÉäÉWÉáÉì	çÇÇ≥íäèo							*/
  /*-------------------------------------------------*/
  void	Collison_put(FIXED x,FIXED y,FIXED z)
  {
--- 492,498 ----
  	/*---------------------------------------------*/
  }
  /*-------------------------------------------------*/
! /*	ÔøΩRÔøΩÔøΩÔøΩWÔøΩÔøΩÔøΩÔøΩ	ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩo							*/
  /*-------------------------------------------------*/
  void	Collison_put(FIXED x,FIXED y,FIXED z)
  {
***************
*** 511,517 ****
  
  	ux	=x>>(23);
  	uy	=y>>(23);
! 	
  	ax_min	=ux-1;
  	ax_max	=ux+2;
  	if(ax_min<0)ax_min=0;
--- 514,520 ----
  
  	ux	=x>>(23);
  	uy	=y>>(23);
! 
  	ax_min	=ux-1;
  	ax_max	=ux+2;
  	if(ax_min<0)ax_min=0;
***************
*** 593,596 ****
  		}
  	}
  }
! 
\ No newline at end of file
--- 596,599 ----
  		}
  	}
  }
! 
diff -crBN --binary sgl302/sample/driving2/main.c sgl302_patch/sample/driving2/main.c
*** sgl302/sample/driving2/main.c	1996-09-10 23:00:00.000000000 -0400
--- sgl302_patch/sample/driving2/main.c	2024-08-11 12:23:22.000000000 -0400
***************
*** 11,16 ****
--- 11,19 ----
  
  #include	"sglcolli.h"
  
+ #include <math.h>
+ #include <stdlib.h>
+ 
  #define FLASH_SATURN_CD
  
  #define	pol_map_x	64
***************
*** 185,199 ****
  /*--------------------------------------------------------------------------*/
  	init_camera((void*)&sega3D);
  	init_object((void*)&sega3D);
! 	init_matrix(pos, ang, scl);	
  	init_main();
  	pos_move();
  	slZdspLevel(5);
  /*--------------------------------------------------------------------------*/
! 	slInitEvent();							/*  ÉCÉxÉìÉgÇÃèâä˙âª		*/
! 	slSetEvent((void *)main_pad);			/*  ÉpÉbÉhì¸óÕä÷åW			*/
! 	slSetEvent((void *)main_came);			/*  ÉJÉÅÉâà íuéwíËä÷åW		*/
! 	slSetEvent((void *)main_Pmap);			/*  É|ÉäÉSÉìÉ}ÉbÉvï\é¶ä÷åW	*/
  /*--------------------------------------------------------------------------*/
  	slPerspective(DEGtoANG(80));
  	slSetDepthLimit(0,11,5);
--- 188,202 ----
  /*--------------------------------------------------------------------------*/
  	init_camera((void*)&sega3D);
  	init_object((void*)&sega3D);
! 	init_matrix(pos, ang, scl);
  	init_main();
  	pos_move();
  	slZdspLevel(5);
  /*--------------------------------------------------------------------------*/
! 	slInitEvent();							/*  ÔøΩCÔøΩxÔøΩÔøΩÔøΩgÔøΩÃèÔøΩÔøΩÔøΩÔøΩÔøΩ		*/
! 	slSetEvent((void *)main_pad);			/*  ÔøΩpÔøΩbÔøΩhÔøΩÔøΩÔøΩÕä÷åW			*/
! 	slSetEvent((void *)main_came);			/*  ÔøΩJÔøΩÔøΩÔøΩÔøΩÔøΩ íuÔøΩwÔøΩÔøΩÔøΩ÷åW		*/
! 	slSetEvent((void *)main_Pmap);			/*  ÔøΩ|ÔøΩÔøΩÔøΩSÔøΩÔøΩÔøΩ}ÔøΩbÔøΩvÔøΩ\ÔøΩÔøΩÔøΩ÷åW	*/
  /*--------------------------------------------------------------------------*/
  	slPerspective(DEGtoANG(80));
  	slSetDepthLimit(0,11,5);
***************
*** 211,225 ****
  			SYS_Exit(1);
  		}
  #endif
! 		slExecuteEvent();					/*	ÉCÉxÉìÉgÇÃé¿çs			*/
  		slPrintFX(TotalPolygons<<16, slLocate(30,4));
  		slPrintFX(DispPolygons<<16, slLocate(30,6));
! 		slSynch();							/*	ÉVÉìÉN					*/
! 	} 
  /*--------------------------------------------------------------------------*/
  }
  /*--------------------------------------*/
! /*	ÉJÉÅÉâä÷åW							*/
  /*--------------------------------------*/
  void	main_came(EVENT *evptr)
  {
--- 214,228 ----
  			SYS_Exit(1);
  		}
  #endif
! 		slExecuteEvent();					/*	ÔøΩCÔøΩxÔøΩÔøΩÔøΩgÔøΩÃéÔøΩÔøΩs			*/
  		slPrintFX(TotalPolygons<<16, slLocate(30,4));
  		slPrintFX(DispPolygons<<16, slLocate(30,6));
! 		slSynch();							/*	ÔøΩVÔøΩÔøΩÔøΩN					*/
! 	}
  /*--------------------------------------------------------------------------*/
  }
  /*--------------------------------------*/
! /*	ÔøΩJÔøΩÔøΩÔøΩÔøΩÔøΩ÷åW							*/
  /*--------------------------------------*/
  void	main_came(EVENT *evptr)
  {
***************
*** 285,291 ****
  	slPopMatrix();
  }
  /*--------------------------------------*/
! /*	à íuï€ë∂							*/
  /*--------------------------------------*/
  void	pos_move()
  {
--- 288,294 ----
  	slPopMatrix();
  }
  /*--------------------------------------*/
! /*	ÔøΩ íuÔøΩ€ëÔøΩ							*/
  /*--------------------------------------*/
  void	pos_move()
  {
***************
*** 297,303 ****
  	ang2[Z]	=ang[Z];
  }
  /*--------------------------------------*/
! /*	ÉpÉbÉhä÷åW							*/
  /*--------------------------------------*/
  void	main_pad(EVENT *evptr)
  {
--- 300,306 ----
  	ang2[Z]	=ang[Z];
  }
  /*--------------------------------------*/
! /*	ÔøΩpÔøΩbÔøΩhÔøΩ÷åW							*/
  /*--------------------------------------*/
  void	main_pad(EVENT *evptr)
  {
***************
*** 307,313 ****
  	Sint16		aa;
  	Uint16		data;
  
! 	data = Smpc_Peripheral[15].data;					//ÉpÉbÉhÇQ
  	if((data & PER_DGT_TC) == 0)tak_data+=10;
  	if((data & PER_DGT_TZ) == 0)tak_data-=10;
  	if((data & PER_DGT_KL) == 0)ang[Z]+=DEGtoANG(4.0);
--- 310,316 ----
  	Sint16		aa;
  	Uint16		data;
  
! 	data = Smpc_Peripheral[15].data;					//ÔøΩpÔøΩbÔøΩhÔøΩQ
  	if((data & PER_DGT_TC) == 0)tak_data+=10;
  	if((data & PER_DGT_TZ) == 0)tak_data-=10;
  	if((data & PER_DGT_KL) == 0)ang[Z]+=DEGtoANG(4.0);
***************
*** 315,321 ****
  	if((data & PER_DGT_KU) == 0)ang[X]+=DEGtoANG(4.0);
  	if((data & PER_DGT_KD) == 0)ang[X]-=DEGtoANG(4.0);
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[0].data;						//ÉpÉbÉhÇP
  #ifdef FLASH_SATURN_CD
  	if((data & (PER_DGT_ST | PER_DGT_TA | PER_DGT_TB | PER_DGT_TC)) == 0){
  		SYS_Exit(0);
--- 318,324 ----
  	if((data & PER_DGT_KU) == 0)ang[X]+=DEGtoANG(4.0);
  	if((data & PER_DGT_KD) == 0)ang[X]-=DEGtoANG(4.0);
  	/*---------------------------------------------*/
! 	data = Smpc_Peripheral[0].data;						//ÔøΩpÔøΩbÔøΩhÔøΩP
  #ifdef FLASH_SATURN_CD
  	if((data & (PER_DGT_ST | PER_DGT_TA | PER_DGT_TB | PER_DGT_TC)) == 0){
  		SYS_Exit(0);
***************
*** 323,331 ****
  #endif
  	if((data & PER_DGT_ST) == 0)init_main();
  /*---------------------------------------------*/
! /*é‘ÇÃãììÆÇÕÉAÉoÉEÉgÇ…çÏÇ¡ÇƒÇ¢ÇÈÇÃÇ≈éQçlÇ…ÇµÇ»Ç¢Ç≈Ç≠ÇæÇ≥Ç¢ÅB*/
  /*---------------------------------------------*/
! 	/*â¡ë¨å∏ë¨*/
  	aa	=0;
  	if((data & PER_DGT_TB) == 0)
  	{
--- 326,334 ----
  #endif
  	if((data & PER_DGT_ST) == 0)init_main();
  /*---------------------------------------------*/
! /*ÔøΩ‘ÇÃãÔøΩÔøΩÔøΩÔøΩÕÉAÔøΩoÔøΩEÔøΩgÔøΩ…çÔøΩÔøΩÔøΩÔøΩƒÇÔøΩÔøΩÔøΩÔøΩÃÇ≈éQÔøΩlÔøΩ…ÇÔøΩÔøΩ»ÇÔøΩÔøΩ≈ÇÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩB*/
  /*---------------------------------------------*/
! 	/*ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩ*/
  	aa	=0;
  	if((data & PER_DGT_TB) == 0)
  	{
***************
*** 359,365 ****
  	pos_add[X]	-=eg_speed*slSin(ang[Z])/10;
  	pos_add[Y]	-=eg_speed*slCos(ang[Z])/10;
  	/*---------------------------------------------*/
! 	/*ÉTÉCÉhà⁄ìÆ*/
  	if((data & PER_DGT_TL) == 0)
  	{
  		pos[X]	+=3*slCos(ang[Z]);
--- 362,368 ----
  	pos_add[X]	-=eg_speed*slSin(ang[Z])/10;
  	pos_add[Y]	-=eg_speed*slCos(ang[Z])/10;
  	/*---------------------------------------------*/
! 	/*ÔøΩTÔøΩCÔøΩhÔøΩ⁄ìÔøΩ*/
  	if((data & PER_DGT_TL) == 0)
  	{
  		pos[X]	+=3*slCos(ang[Z]);
***************
*** 370,376 ****
  		pos[Y]	+=3*slSin(ang[Z]);
  	}
  	/*---------------------------------------------*/
! 	/*ç∂âEÇÃâÒì]*/
  	aa	=0;
  	if((data & PER_DGT_KR) == 0)
  	{
--- 373,379 ----
  		pos[Y]	+=3*slSin(ang[Z]);
  	}
  	/*---------------------------------------------*/
! 	/*ÔøΩÔøΩÔøΩEÔøΩÃâÔøΩÔøΩ]*/
  	aa	=0;
  	if((data & PER_DGT_KR) == 0)
  	{
***************
*** 418,424 ****
  	by	=ay-30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
  	point_l	=col_att;
! 	
  	bx	=ax-30*slCos(ang[Z]);
  	by	=ay+30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
--- 421,427 ----
  	by	=ay-30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
  	point_l	=col_att;
! 
  	bx	=ax-30*slCos(ang[Z]);
  	by	=ay+30*slSin(ang[Z]);
  	Collison_put((bx),(by),pos[Z]);
***************
*** 475,481 ****
  	/*---------------------------------------------*/
  }
  /*-------------------------------------------------*/
! /*	ÉRÉäÉWÉáÉì	çÇÇ≥íäèo							*/
  /*-------------------------------------------------*/
  void	Collison_put(FIXED x,FIXED y,FIXED z)
  {
--- 478,484 ----
  	/*---------------------------------------------*/
  }
  /*-------------------------------------------------*/
! /*	ÔøΩRÔøΩÔøΩÔøΩWÔøΩÔøΩÔøΩÔøΩ	ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩo							*/
  /*-------------------------------------------------*/
  void	Collison_put(FIXED x,FIXED y,FIXED z)
  {
***************
*** 495,501 ****
  
  	ux	=x>>(23);
  	uy	=y>>(23);
! 	
  	ax_min	=ux-1;
  	ax_max	=ux+2;
  	if(ax_min<0)ax_min=0;
--- 498,504 ----
  
  	ux	=x>>(23);
  	uy	=y>>(23);
! 
  	ax_min	=ux-1;
  	ax_max	=ux+2;
  	if(ax_min<0)ax_min=0;
***************
*** 559,562 ****
  		col_grp		=GetCol[GetColNo].gru;
  	}
  }
! 
\ No newline at end of file
--- 562,565 ----
  		col_grp		=GetCol[GetColNo].gru;
  	}
  }
! 
diff -crBN --binary sgl302/sample/s_10/main.c sgl302_patch/sample/s_10/main.c
*** sgl302/sample/s_10/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/s_10/main.c	2024-08-11 12:44:00.000000000 -0400
***************
*** 3,8 ****
--- 3,10 ----
  /*----------------------------------------------------------------------*/
  #include	"sgl.h"
  
+ void set_event();
+ 
  void ss_main(void)
  {
  	slInitSystem(TV_320x224,NULL,1) ;
***************
*** 14,17 ****
  		slSynch() ;
  	}
  }
! 
\ No newline at end of file
--- 16,19 ----
  		slSynch() ;
  	}
  }
! 
diff -crBN --binary sgl302/sample/s_4_2/main.c sgl302_patch/sample/s_4_2/main.c
*** sgl302/sample/s_4_2/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/s_4_2/main.c	2024-08-11 12:39:04.000000000 -0400
***************
*** 7,13 ****
  
  void ss_main(void){
  	static ANGLE	ang[XYZ];
- 	static ANGLE	angper[XYZ];
  	static FIXED	pos[XYZ];
  	static ANGLE	tmp = DEGtoANG(90.0);
  	static ANGLE	adang = DEGtoANG(0.5);
--- 7,12 ----
***************
*** 28,34 ****
  		slPrintHex(slAng2Dec(tmp) , slLocate(18,5));
  
  		tmp += adang;
! 		if (tmp > DEGtoANG(160.0)) 
  			adang = DEGtoANG(-0.5);
  		if (tmp < DEGtoANG(20.0))
  			adang = DEGtoANG(0.5);
--- 27,33 ----
  		slPrintHex(slAng2Dec(tmp) , slLocate(18,5));
  
  		tmp += adang;
! 		if (tmp > DEGtoANG(160.0))
  			adang = DEGtoANG(-0.5);
  		if (tmp < DEGtoANG(20.0))
  			adang = DEGtoANG(0.5);
***************
*** 49,52 ****
  		slSynch();
  	}
  }
! 
\ No newline at end of file
--- 48,51 ----
  		slSynch();
  	}
  }
! 
diff -crBN --binary sgl302/sample/s_6_3/main.c sgl302_patch/sample/s_6_3/main.c
*** sgl302/sample/s_6_3/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/s_6_3/main.c	2024-08-11 12:38:34.000000000 -0400
***************
*** 32,38 ****
  
  void ss_main(void)
  {
- 	static ANGLE	ang[XYZ];
  	static CAMERA	cmbuf;
  	static ANGLE	tmp = DEGtoANG(0.0);
  
--- 32,37 ----
***************
*** 62,65 ****
  	}
  }
  
! 
\ No newline at end of file
--- 61,64 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample/s_8_9_1/main.c sgl302_patch/sample/s_8_9_1/main.c
*** sgl302/sample/s_8_9_1/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/s_8_9_1/main.c	2024-08-11 12:37:36.000000000 -0400
***************
*** 13,19 ****
  void ss_main(void)
  {
  	ANGLE yama_angz = DEGtoANG(0.0);
- 	FIXED posx = toFIXED(128.0) , posy = toFIXED(64.0);
  	
  	slInitSystem(TV_320x224,NULL,1);
  	slTVOff();
--- 13,18 ----
***************
*** 42,48 ****
  		slZrotR(yama_angz);
  		yama_angz += DEGtoANG(1.0);
  		slSynch();
! 	} 
  }
  
! 
\ No newline at end of file
--- 41,47 ----
  		slZrotR(yama_angz);
  		yama_angz += DEGtoANG(1.0);
  		slSynch();
! 	}
  }
  
! 
diff -crBN --binary sgl302/sample/s_9_1/main.c sgl302_patch/sample/s_9_1/main.c
*** sgl302/sample/s_9_1/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/s_9_1/main.c	2024-08-11 12:36:20.000000000 -0400
***************
*** 3,8 ****
--- 3,9 ----
  /*----------------------------------------------------------------------*/
  #include	"sgl.h"
  #include	"sega_sys.h"
+ #include	"ss_scrol.h"
  
  #define		NBG1_CEL_ADR		( VDP2_VRAM_B1 + 0x02000 )
  #define		NBG1_MAP_ADR		( VDP2_VRAM_B1 + 0x12000 )
***************
*** 26,34 ****
  	PER_DGT_TL,
  };
  
! extern pad_cel[];
! extern pad_map[];
! extern pad_pal[];
  extern TEXTURE tex_spr[];
  extern PICTURE pic_spr[];
  extern FIXED stat[][XYZS];
--- 27,35 ----
  	PER_DGT_TL,
  };
  
! extern Uint8 pad_cel[];
! extern Uint16 pad_map[];
! extern Uint16 pad_pal[];
  extern TEXTURE tex_spr[];
  extern PICTURE pic_spr[];
  extern FIXED stat[][XYZS];
***************
*** 38,44 ****
  static void set_sprite(PICTURE *pcptr , Uint32 NbPicture)
  {
  	TEXTURE *txptr;
!  
  	for(; NbPicture-- > 0; pcptr++){
  		txptr = tex_spr + pcptr->texno;
  		slDMACopy((void *)pcptr->pcsrc,
--- 39,45 ----
  static void set_sprite(PICTURE *pcptr , Uint32 NbPicture)
  {
  	TEXTURE *txptr;
! 
  	for(; NbPicture-- > 0; pcptr++){
  		txptr = tex_spr + pcptr->texno;
  		slDMACopy((void *)pcptr->pcsrc,
***************
*** 69,75 ****
  	slTVOff();
  	set_sprite(pic_spr,1);
  	slPrint("Sample program 9.1" , slLocate(9,2));
! 	
  	slColRAMMode(CRM16_1024);
  	slBack1ColSet((void *)BACK_COL_ADR , 0);
  
--- 70,76 ----
  	slTVOff();
  	set_sprite(pic_spr,1);
  	slPrint("Sample program 9.1" , slLocate(9,2));
! 
  	slColRAMMode(CRM16_1024);
  	slBack1ColSet((void *)BACK_COL_ADR , 0);
  
***************
*** 84,94 ****
  	slScrPosNbg1(toFIXED(-32.0) , toFIXED(-36.0));
  	slScrAutoDisp(NBG0ON | NBG1ON);
  	slTVOn();
! 	
  	while(1) {
  		disp_sprite();
  		slSynch();
! 	} 
  }
  
! 
\ No newline at end of file
--- 85,95 ----
  	slScrPosNbg1(toFIXED(-32.0) , toFIXED(-36.0));
  	slScrAutoDisp(NBG0ON | NBG1ON);
  	slTVOn();
! 
  	while(1) {
  		disp_sprite();
  		slSynch();
! 	}
  }
  
! 
diff -crBN --binary sgl302/sample/sbklib/string.c sgl302_patch/sample/sbklib/string.c
*** sgl302/sample/sbklib/string.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample/sbklib/string.c	2022-09-15 21:12:04.000000000 -0400
***************
*** 158,164 ****
      pt = s;
  
      for( i=0; i<n; i++ )
!       *( ((char *)pt)++ ) = *( ( (char *)ct )++ );
  
      return (char *)s;
  }
--- 158,164 ----
      pt = s;
  
      for( i=0; i<n; i++ )
!       *((char *)(pt + i)) = *((char *)(ct + i));
  
      return (char *)s;
  }
***************
*** 168,186 ****
      void	*pt;
      int		i;
  
-     pt = s;
      if ( (int)s < (int)ct )
        {
  	  for( i=0; i<n; i++ )
! 	    *( ((char *)pt)++ ) = *( ( (char *)ct )++ );
  
        }
      else
        {
! 	  (char *)pt += n - 1;
! 	  (char *)ct += n - 1;
  	  for( i=0; i<n; i++ )
! 	    *( ((char *)pt)-- ) = *( ( (char *)ct )-- );
        }
  
  
--- 168,187 ----
      void	*pt;
      int		i;
  
      if ( (int)s < (int)ct )
        {
+           pt = s;
+           
  	  for( i=0; i<n; i++ )
! 	    *((char *)(pt + i)) = *((char *)(ct + i));
  
        }
      else
        {
! 	  pt = s + n - 1;
! 	  void	*pct = ct + n - 1;
  	  for( i=0; i<n; i++ )
! 	    *((char *)(pt - i)) = *((char *)(pct - i));
        }
  
  
***************
*** 195,201 ****
      pt = s;
      for(i=0; i<n; i++)
        {
! 	  *((char *)pt)++ = c;
        }
  
      return s;
--- 196,202 ----
      pt = s;
      for(i=0; i<n; i++)
        {
! 	  *((char *)(pt + i)) = c;
        }
  
      return s;
diff -crBN --binary sgl302/sample2/common/cinit.c sgl302_patch/sample2/common/cinit.c
*** sgl302/sample2/common/cinit.c	1996-09-10 23:00:00.000000000 -0400
--- sgl302_patch/sample2/common/cinit.c	2022-09-15 21:13:10.000000000 -0400
***************
*** 14,20 ****
  /* */
  extern void ss_main( void );
  
! void	main( void )
  {
  	Uint8	*dst;
  	Uint32	i;
--- 14,20 ----
  /* */
  extern void ss_main( void );
  
! int	main( void )
  {
  	Uint8	*dst;
  	Uint32	i;
***************
*** 32,36 ****
  
  	/* Application Call */
  	ss_main();
  }
! 
\ No newline at end of file
--- 32,38 ----
  
  	/* Application Call */
  	ss_main();
+ 	
+ 	return 0;
  }
! 
diff -crBN --binary sgl302/sample2/sega2d_1/scroll.c sgl302_patch/sample2/sega2d_1/scroll.c
*** sgl302/sample2/sega2d_1/scroll.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega2d_1/scroll.c	2022-09-15 19:13:48.000000000 -0400
***************
*** 1,5 ****
  #include	"sgl.h"
! #include	"sega2D.h"
  
  #include	SEGA2D_MAP
  #include	SEGA2D_CEL
--- 1,5 ----
  #include	"sgl.h"
! #include	"sega2d.h"
  
  #include	SEGA2D_MAP
  #include	SEGA2D_CEL
***************
*** 7,10 ****
  
  
  
! 
\ No newline at end of file
--- 7,10 ----
  
  
  
! 
diff -crBN --binary sgl302/sample2/sega2d_1/sega2d.h sgl302_patch/sample2/sega2d_1/sega2d.h
*** sgl302/sample2/sega2d_1/sega2d.h	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega2d_1/sega2d.h	2022-09-15 18:56:10.000000000 -0400
***************
*** 1,9 ****
  
! #define	SEGA2D_MAP	".\scroll\vf.map"
! #define	SEGA2D_CEL	".\scroll\vf.cel"
! #define	SEGA2D_PAL	".\scroll\vf.pal"
  
! #define	SEGA2D_H	".\scroll\vf.h"
  
  
! 
\ No newline at end of file
--- 1,9 ----
  
! #define	SEGA2D_MAP	"./scroll/vf.map"
! #define	SEGA2D_CEL	"./scroll/vf.cel"
! #define	SEGA2D_PAL	"./scroll/vf.pal"
  
! #define	SEGA2D_H	"./scroll/vf.h"
  
  
! 
diff -crBN --binary sgl302/sample2/sega3d_1/3d/dolphin.c sgl302_patch/sample2/sega3d_1/3d/dolphin.c
*** sgl302/sample2/sega3d_1/3d/dolphin.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_1/3d/dolphin.c	2022-09-15 19:22:56.000000000 -0400
***************
*** 1,4 ****
  
! #include	".\3d\model_2.c"
! #include	".\3d\model_4.c"
  
--- 1,4 ----
  
! #include	"./3d/model_2.c"
! #include	"./3d/model_4.c"
  
diff -crBN --binary sgl302/sample2/sega3d_1/main.c sgl302_patch/sample2/sega3d_1/main.c
*** sgl302/sample2/sega3d_1/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_1/main.c	2024-08-02 20:54:20.000000000 -0400
***************
*** 17,22 ****
--- 17,23 ----
  	FIXED	scl[XYZ];
  }Sega3D, *Sega3DPtr;
  
+ void	chg_view(void *ptr);
  
  static void set_texture(void)
  {
***************
*** 130,133 ****
  	}
  }
  
! 
\ No newline at end of file
--- 131,134 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample2/sega3d_1/sega3d.h sgl302_patch/sample2/sega3d_1/sega3d.h
*** sgl302/sample2/sega3d_1/sega3d.h	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_1/sega3d.h	2022-09-15 18:57:10.000000000 -0400
***************
*** 1,13 ****
  
! #define	SEGA3D_INC		"3d\dolphin.h"
! #define	SEGA3D_INI		"3d\dolphin.ini"
! #define	SEGA3D_CC		"3d\dolphin.cc"
! #define	SEGA3D_C		"3d\dolphin.c"
! #define	SEGA3D_GR		"3d\dolphin.gr"
! #define	SEGA3D_LIGHT	"3d\light.ini"
  
! #define	SEGA3D_TEXDEF	"texture\texture.def"
! #define	SEGA3D_TEXC		"texture\dolphin.c"
  #define	SEGA3D_TEX		tex_dolphin
  #define	SEGA3D_PIC		pic_dolphin
  #define	SEGA3D_MAXTEX	96
--- 1,13 ----
  
! #define	SEGA3D_INC		"3d/dolphin.h"
! #define	SEGA3D_INI		"3d/dolphin.ini"
! #define	SEGA3D_CC		"3d/dolphin.cc"
! #define	SEGA3D_C		"3d/dolphin.c"
! #define	SEGA3D_GR		"3d/dolphin.gr"
! #define	SEGA3D_LIGHT	"3d/light.ini"
  
! #define	SEGA3D_TEXDEF	"texture/texture.def"
! #define	SEGA3D_TEXC		"texture/dolphin.c"
  #define	SEGA3D_TEX		tex_dolphin
  #define	SEGA3D_PIC		pic_dolphin
  #define	SEGA3D_MAXTEX	96
***************
*** 28,31 ****
  #define	VTXROT90(a,b,c,d)	VERTICES(b,c,d,a)	/*0123->1230*/
  #endif
  
! 
\ No newline at end of file
--- 28,31 ----
  #define	VTXROT90(a,b,c,d)	VERTICES(b,c,d,a)	/*0123->1230*/
  #endif
  
! 
diff -crBN --binary sgl302/sample2/sega3d_2/3d/cube.c sgl302_patch/sample2/sega3d_2/3d/cube.c
*** sgl302/sample2/sega3d_2/3d/cube.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_2/3d/cube.c	2022-09-15 19:38:04.000000000 -0400
***************
*** 1,3 ****
  
! #include	".\3d\new_1.c"
  
--- 1,3 ----
  
! #include	"./3d/new_1.c"
  
diff -crBN --binary sgl302/sample2/sega3d_2/main.c sgl302_patch/sample2/sega3d_2/main.c
*** sgl302/sample2/sega3d_2/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_2/main.c	2024-08-02 21:04:38.000000000 -0400
***************
*** 19,24 ****
--- 19,25 ----
  	FIXED	scl[XYZ];
  }Sega3D, *Sega3DPtr;
  
+ void	chg_view(void *ptr);
  
  static void set_texture(void)
  {
***************
*** 136,139 ****
  	}
  }
  
! 
\ No newline at end of file
--- 137,140 ----
  	}
  }
  
! 
diff -crBN --binary sgl302/sample2/sega3d_2/sega3d.h sgl302_patch/sample2/sega3d_2/sega3d.h
*** sgl302/sample2/sega3d_2/sega3d.h	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_2/sega3d.h	2022-09-15 19:28:20.000000000 -0400
***************
*** 1,12 ****
! #define	SEGA3D_INC		"3d\cube.h"
! #define	SEGA3D_INI		"3d\cube.ini"
! #define	SEGA3D_CC		"3d\cube.cc"
! #define	SEGA3D_C		"3d\cube.c"
! #define	SEGA3D_GR		"3d\cube.gr"
! #define	SEGA3D_LIGHT	"3d\light.ini"
  
! #define	SEGA3D_TEXDEF	"texture\texture.def"
! #define	SEGA3D_TEXC		"texture\cube.c"
  #define	SEGA3D_TEX		tex_cube
  #define	SEGA3D_PIC		pic_cube
  #define	SEGA3D_MAXTEX	3
--- 1,12 ----
! #define	SEGA3D_INC		"3d/cube.h"
! #define	SEGA3D_INI		"3d/cube.ini"
! #define	SEGA3D_CC		"3d/cube.cc"
! #define	SEGA3D_C		"3d/cube.c"
! #define	SEGA3D_GR		"3d/cube.gr"
! #define	SEGA3D_LIGHT	"3d/light.ini"
  
! #define	SEGA3D_TEXDEF	"texture/texture.def"
! #define	SEGA3D_TEXC		"texture/cube.c"
  #define	SEGA3D_TEX		tex_cube
  #define	SEGA3D_PIC		pic_cube
  #define	SEGA3D_MAXTEX	3
***************
*** 29,32 ****
  #define	VTXROT90(a,b,c,d)	VERTICES(b,c,d,a)	/*0123->1230*/
  #endif
  
! 
\ No newline at end of file
--- 29,32 ----
  #define	VTXROT90(a,b,c,d)	VERTICES(b,c,d,a)	/*0123->1230*/
  #endif
  
! 
diff -crBN --binary sgl302/sample2/sega3d_2/texture/cube.c sgl302_patch/sample2/sega3d_2/texture/cube.c
*** sgl302/sample2/sega3d_2/texture/cube.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_2/texture/cube.c	2022-09-15 19:28:52.000000000 -0400
***************
*** 1,6 ****
! #include	".\texture\saturn.c"
! #include	".\texture\tails.c"
! #include	".\texture\mika.c"
  
  TEXTURE	tex_cube[] = {
  	{320, 82, 0x2000, 0x2852},
--- 1,6 ----
! #include	"./texture/saturn.c"
! #include	"./texture/tails.c"
! #include	"./texture/mika.c"
  
  TEXTURE	tex_cube[] = {
  	{320, 82, 0x2000, 0x2852},
diff -crBN --binary sgl302/sample2/sega3d_3/3d/opa.c sgl302_patch/sample2/sega3d_3/3d/opa.c
*** sgl302/sample2/sega3d_3/3d/opa.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/3d/opa.c	2022-09-15 20:18:12.000000000 -0400
***************
*** 1,8 ****
  
! #include	".\3d\model_3.c"
! #include	".\3d\model_5.c"
! #include	".\3d\model_7.c"
! #include	".\3d\model_9.c"
! #include	".\3d\model_11.c"
! #include	".\3d\model_13.c"
  
--- 1,8 ----
  
! #include	"./3d/model_3.c"
! #include	"./3d/model_5.c"
! #include	"./3d/model_7.c"
! #include	"./3d/model_9.c"
! #include	"./3d/model_11.c"
! #include	"./3d/model_13.c"
  
diff -crBN --binary sgl302/sample2/sega3d_3/action.c sgl302_patch/sample2/sega3d_3/action.c
*** sgl302/sample2/sega3d_3/action.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/action.c	2022-09-15 19:33:12.000000000 -0400
***************
*** 1,7 ****
  #include	"sgl.h"
  #include	"sega_sys.h"
  
! #include	"sega3D.h"
  #include	"scsp.h"
  
  
--- 1,7 ----
  #include	"sgl.h"
  #include	"sega_sys.h"
  
! #include	"sega3d.h"
  #include	"scsp.h"
  
  
***************
*** 217,220 ****
  
  
  
! 
\ No newline at end of file
--- 217,220 ----
  
  
  
! 
diff -crBN --binary sgl302/sample2/sega3d_3/main.c sgl302_patch/sample2/sega3d_3/main.c
*** sgl302/sample2/sega3d_3/main.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/main.c	2024-08-05 15:57:20.000000000 -0400
***************
*** 27,32 ****
--- 27,34 ----
  static void set_matrix(FIXED *, ANGLE *, FIXED *);
  static void init_matrix(FIXED *, ANGLE *, FIXED *);
  
+ void			whichAction(void *);
+ 
  void	init_camera(void *ptr)
  {
  	Sega3DPtr	segaPtr = (Sega3DPtr)ptr;
***************
*** 110,116 ****
  #endif
  }
  
! static setAngLimit(ANGLE *ang, float min, float max)
  {
  	if(*ang > DEGtoANG(max))
  		*ang = DEGtoANG(max);
--- 112,118 ----
  #endif
  }
  
! static void setAngLimit(ANGLE *ang, float min, float max)
  {
  	if(*ang > DEGtoANG(max))
  		*ang = DEGtoANG(max);
***************
*** 226,232 ****
  		reset_vbar(39);
  		slSynch();
  		set_vbar(39);
! 	} 
  }
  
! 
\ No newline at end of file
--- 228,234 ----
  		reset_vbar(39);
  		slSynch();
  		set_vbar(39);
! 	}
  }
  
! 
diff -crBN --binary sgl302/sample2/sega3d_3/scroll.c sgl302_patch/sample2/sega3d_3/scroll.c
*** sgl302/sample2/sega3d_3/scroll.c	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/scroll.c	2022-09-15 20:24:12.000000000 -0400
***************
*** 1,5 ****
  #include	"sgl.h"
! #include	"sega2D.h"
  
  #include	SEGA2D_MAP
  #include	SEGA2D_CEL
--- 1,5 ----
  #include	"sgl.h"
! #include	"sega2d.h"
  
  #include	SEGA2D_MAP
  #include	SEGA2D_CEL
***************
*** 7,10 ****
  
  
  
! 
\ No newline at end of file
--- 7,10 ----
  
  
  
! 
diff -crBN --binary sgl302/sample2/sega3d_3/sega2d.h sgl302_patch/sample2/sega3d_3/sega2d.h
*** sgl302/sample2/sega3d_3/sega2d.h	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/sega2d.h	2022-09-15 20:19:00.000000000 -0400
***************
*** 1,9 ****
  
! #define	SEGA2D_MAP	".\scroll\vf.map"
! #define	SEGA2D_CEL	".\scroll\vf.cel"
! #define	SEGA2D_PAL	".\scroll\vf.pal"
  
! #define	SEGA2D_H	".\scroll\vf.h"
  
  
! 
\ No newline at end of file
--- 1,9 ----
  
! #define	SEGA2D_MAP	"./scroll/vf.map"
! #define	SEGA2D_CEL	"./scroll/vf.cel"
! #define	SEGA2D_PAL	"./scroll/vf.pal"
  
! #define	SEGA2D_H	"./scroll/vf.h"
  
  
! 
diff -crBN --binary sgl302/sample2/sega3d_3/sega3d.h sgl302_patch/sample2/sega3d_3/sega3d.h
*** sgl302/sample2/sega3d_3/sega3d.h	1996-04-20 22:10:00.000000000 -0400
--- sgl302_patch/sample2/sega3d_3/sega3d.h	2022-09-15 20:19:28.000000000 -0400
***************
*** 1,13 ****
  
! #define	SEGA3D_H		"3d\opa.h"
! #define	SEGA3D_INI		"3d\opa.ini"
! #define	SEGA3D_CC		"3d\opa.cc"
! #define	SEGA3D_C		"3d\opa.c"
! #define	SEGA3D_GR		"3d\opa.gr"
! #define	SEGA3D_LIGHT	"3d\light.ini"
  
! #define	SEGA3D_TEXDEF	"texture\texture.def"
! #define	SEGA3D_TEXC		"texture\opa.c"
  #define	SEGA3D_TEX		tex_opa
  #define	SEGA3D_PIC		pic_opa
  #define	SEGA3D_MAXTEX	1
--- 1,13 ----
  
! #define	SEGA3D_H		"3d/opa.h"
! #define	SEGA3D_INI		"3d/opa.ini"
! #define	SEGA3D_CC		"3d/opa.cc"
! #define	SEGA3D_C		"3d/opa.c"
! #define	SEGA3D_GR		"3d/opa.gr"
! #define	SEGA3D_LIGHT	"3d/light.ini"
  
! #define	SEGA3D_TEXDEF	"texture/texture.def"
! #define	SEGA3D_TEXC		"texture/opa.c"
  #define	SEGA3D_TEX		tex_opa
  #define	SEGA3D_PIC		pic_opa
  #define	SEGA3D_MAXTEX	1
***************
*** 44,47 ****
  #endif
  
  
! 
\ No newline at end of file
--- 44,47 ----
  #endif
  
  
! 
